--- a/arch/arm/mach-pentagram/sp7021.c	2020-08-23 02:13:47.153688363 +0300
+++ a/arch/arm/mach-pentagram/sp7021.c	2020-08-23 20:14:29.563069978 +0300
@@ -40,8 +40,8 @@
 	//printk("PD RG_PLL_PDN and RG_PLLIO_PDN to save power\n");
 	writel(0, regs + 0x54); /* bit0 RG_PLLIO_PDN */
 	writel(0, regs + 0x2C); /* bit0 RG_PLL_PDN */
-	while (1)
-		;
+//	while (1)
+//		;
 
 //	printk("PD Achip mo_gclk_en0/mo_clk_en0 to save power \n");
 //	writel(0, regs + 0x28); 
