--- a/arch/arm/mach-pentagram/sp7021.c	2020-08-13 14:40:15.937079660 +0300
+++ a/arch/arm/mach-pentagram/sp7021.c	2020-08-14 13:21:38.073792392 +0300
@@ -40,6 +40,8 @@
 	//printk("PD RG_PLL_PDN and RG_PLLIO_PDN to save power\n");
 	writel(0, regs + 0x54); /* bit0 RG_PLLIO_PDN */
 	writel(0, regs + 0x2C); /* bit0 RG_PLL_PDN */
+	while (1)
+	;
 
 //	printk("PD Achip mo_gclk_en0/mo_clk_en0 to save power \n");
 //	writel(0, regs + 0x28); 
