--- a/drivers/usb/gadget/composite.c	2017-09-20 09:25:38.000000000 +0300
+++ a/drivers/usb/gadget/composite.c	2018-11-12 06:42:55.000000000 +0300
@@ -2003,7 +2003,7 @@
 }
 
 /*-------------------------------------------------------------------------*/
-
+#if 0
 static ssize_t suspended_show(struct device *dev, struct device_attribute *attr,
 			      char *buf)
 {
@@ -2013,7 +2013,7 @@
 	return sprintf(buf, "%d\n", cdev->suspended);
 }
 static DEVICE_ATTR_RO(suspended);
-
+#endif
 static void __composite_unbind(struct usb_gadget *gadget, bool unbind_driver)
 {
 	struct usb_composite_dev	*cdev = get_gadget_data(gadget);
@@ -2098,11 +2098,11 @@
 	cdev->req->buf = kmalloc(USB_COMP_EP0_BUFSIZ, GFP_KERNEL);
 	if (!cdev->req->buf)
 		goto fail;
-
+#if 0
 	ret = device_create_file(&gadget->dev, &dev_attr_suspended);
 	if (ret)
 		goto fail_dev;
-
+#endif
 	cdev->req->complete = composite_setup_complete;
 	cdev->req->context = cdev;
 	gadget->ep0->driver_data = cdev;
@@ -2123,7 +2123,7 @@
 	 */
 	usb_ep_autoconfig_reset(gadget);
 	return 0;
-fail_dev:
+
 	kfree(cdev->req->buf);
 fail:
 	usb_ep_free_request(gadget->ep0, cdev->req);
@@ -2178,7 +2178,7 @@
 		usb_ep_free_request(cdev->gadget->ep0, cdev->req);
 	}
 	cdev->next_string_id = 0;
-	device_remove_file(&cdev->gadget->dev, &dev_attr_suspended);
+	//device_remove_file(&cdev->gadget->dev, &dev_attr_suspended);
 }
 
 static int composite_bind(struct usb_gadget *gadget,
