--- a/drivers/usb/gadget/u_f.h	2017-09-20 09:25:38.000000000 +0300
+++ a/drivers/usb/gadget/u_f.h	2018-11-12 06:42:55.000000000 +0300
@@ -59,13 +59,5 @@
  * usb_requests's length (req->length) to refer to the allocated buffer size.
  * Requests allocated via alloc_ep_req() *must* be freed by free_ep_req().
  */
-struct usb_request *alloc_ep_req(struct usb_ep *ep, size_t len);
-
-/* Frees a usb_request previously allocated by alloc_ep_req() */
-static inline void free_ep_req(struct usb_ep *ep, struct usb_request *req)
-{
-	kfree(req->buf);
-	usb_ep_free_request(ep, req);
-}
 
 #endif /* __U_F_H__ */
