--- a/drivers/usb/gadget/Makefile	2017-09-20 09:25:38.000000000 +0300
+++ a/drivers/usb/gadget/Makefile	2018-11-12 06:42:55.000000000 +0300
@@ -6,7 +6,7 @@
 ccflags-y				+= -I$(srctree)/drivers/usb/gadget/udc
 
 obj-$(CONFIG_USB_LIBCOMPOSITE)	+= libcomposite.o
-libcomposite-y			:= usbstring.o config.o epautoconf.o
+#libcomposite-y			:= usbstring.o config.o epautoconf.o
 libcomposite-y			+= composite.o functions.o configfs.o u_f.o
 
-obj-$(CONFIG_USB_GADGET)	+= udc/ function/ legacy/
+obj-$(CONFIG_USB_GADGET)	+= udc/ legacy/
