--- a/arch/arm/boot/dts/pentagram-sp7021-common.dtsi	2019-10-09 15:17:00.061754084 +0300
+++ a/arch/arm/boot/dts/pentagram-sp7021-common.dtsi	2019-10-09 15:30:19.964838705 +0300
@@ -786,7 +786,7 @@
 			no-sdio;
 			no-mmc;
 		};
-/*
+/* */
 		sdio: sdio@sdcard {
 			compatible = "sunplus,sp7021-sdio";
 			reg = <0x9c008400 0x280>;
@@ -804,7 +804,7 @@
 			no-sd;
 			no-mmc;
 		};
-*/
+/* */
 		spi_controller0: spi@sp_spi_controller0 {
 			compatible = "sunplus,sp7021-spi-controller";
 			reg = <0x9C002D80 0x80>,<0x9C002E00 0x80>;
