--- a/arch/arm/Kconfig	2017-09-20 09:25:38.000000000 +0300
+++ a/arch/arm/Kconfig	2018-11-12 06:42:50.000000000 +0300
@@ -639,8 +639,23 @@
 	help
 	  Support for older TI OMAP1 (omap7xx, omap15xx or omap16xx)
 
+config ARCH_PENTAGRAM
+	bool "Sunplus Pentagram architectures"
+	select CLKSRC_OF
+	select COMMON_CLK
+	select GENERIC_CLOCKEVENTS
+	select GENERIC_IRQ_CHIP
+	select MULTI_IRQ_HANDLER
+	select SERIAL_CORE
+	select SERIAL_CORE_CONSOLE
+	select SERIAL_SP_UART
+	select USE_OF
+	select RTC_CLASS
+
 endchoice
 
+source "arch/arm/mach-pentagram/Kconfig"
+
 menu "Multiple platform selection"
 	depends on ARCH_MULTIPLATFORM
 
