--- a/drivers/usb/host/Makefile	2017-09-20 09:25:38.000000000 +0300
+++ a/drivers/usb/host/Makefile	2018-11-12 06:42:55.000000000 +0300
@@ -3,6 +3,7 @@
 #
 
 # tell define_trace.h where to find the xhci trace header
+ccflags-$(CONFIG_USB_DEBUG) := -DDEBUG
 CFLAGS_xhci-trace.o := -I$(src)
 
 fhci-y := fhci-hcd.o fhci-hub.o fhci-q.o
@@ -31,7 +32,7 @@
 
 obj-$(CONFIG_USB_EHCI_HCD)	+= ehci-hcd.o
 obj-$(CONFIG_USB_EHCI_PCI)	+= ehci-pci.o
-obj-$(CONFIG_USB_EHCI_HCD_PLATFORM)	+= ehci-platform.o
+#obj-$(CONFIG_USB_EHCI_HCD_PLATFORM)	+= ehci-platform.o
 obj-$(CONFIG_USB_EHCI_MXC)	+= ehci-mxc.o
 obj-$(CONFIG_USB_EHCI_HCD_OMAP)	+= ehci-omap.o
 obj-$(CONFIG_USB_EHCI_HCD_ORION)	+= ehci-orion.o
@@ -49,7 +50,7 @@
 
 obj-$(CONFIG_USB_OHCI_HCD)	+= ohci-hcd.o
 obj-$(CONFIG_USB_OHCI_HCD_PCI)	+= ohci-pci.o
-obj-$(CONFIG_USB_OHCI_HCD_PLATFORM)	+= ohci-platform.o
+#obj-$(CONFIG_USB_OHCI_HCD_PLATFORM)	+= ohci-platform.o
 obj-$(CONFIG_USB_OHCI_EXYNOS)	+= ohci-exynos.o
 obj-$(CONFIG_USB_OHCI_HCD_OMAP1)	+= ohci-omap.o
 obj-$(CONFIG_USB_OHCI_HCD_OMAP3)	+= ohci-omap3.o
@@ -81,3 +82,7 @@
 obj-$(CONFIG_USB_HCD_SSB)	+= ssb-hcd.o
 obj-$(CONFIG_USB_FOTG210_HCD)	+= fotg210-hcd.o
 obj-$(CONFIG_USB_MAX3421_HCD)	+= max3421-hcd.o
+obj-$(CONFIG_USB_EHCI_HCD)	+= ehci0-sunplus.o
+obj-$(CONFIG_USB_EHCI_HCD)	+= ehci1-sunplus.o
+obj-$(CONFIG_USB_OHCI_HCD)	+= ohci0-sunplus.o
+obj-$(CONFIG_USB_OHCI_HCD)	+= ohci1-sunplus.o
