--- a/drivers/usb/host/Kconfig	2017-09-20 09:25:38.000000000 +0300
+++ a/drivers/usb/host/Kconfig	2018-11-12 06:42:55.000000000 +0300
@@ -3,6 +3,25 @@
 #
 comment "USB Host Controller Drivers"
 
+config USB_LOGO_TEST
+	bool "GEMINI USB Logo Test support"
+	depends on USB
+	select USB_BAD_DEVICE_INFO
+
+# workaround hw bug that randomly happen abnormal when host controller send packet
+config USB_HOST_RESET_SP
+	bool "Enable Reset SUNPLUS usb host controllor with wait_event"
+	depends on USB
+
+config USB_OHCI_MSI_WORKAROUND
+	bool "Enable Workaround USB Ohci Msi Problem"
+	depends on USB
+
+config USB_USE_PLATFORM_RESOURCE
+	bool "GEMINI Use USB Platfrom Resource"
+	depends on USB
+	default n
+
 config USB_C67X00_HCD
 	tristate "Cypress C67x00 HCD support"
 	depends on HAS_IOMEM
@@ -152,6 +171,20 @@
 		Enables support for the onchip USB controller on the PMC_MSP7100 Family SoC's.
 		If unsure, say N.
 
+config USB_EHCI_BIG_ENDIAN_MMIO
+	bool
+	depends on USB_EHCI_HCD && (PPC_CELLEB || PPC_PS3 || 440EPX || \
+				    ARCH_IXP4XX || XPS_USB_HCD_XILINX || \
+				    PPC_MPC512x || CPU_CAVIUM_OCTEON || \
+				    PMC_MSP || SPARC_LEON)
+	default y
+
+config USB_EHCI_BIG_ENDIAN_DESC
+	bool
+	depends on USB_EHCI_HCD && (440EPX || ARCH_IXP4XX || XPS_USB_HCD_XILINX || \
+				    PPC_MPC512x || PMC_MSP || SPARC_LEON)
+	default y
+
 config XPS_USB_HCD_XILINX
 	bool "Use Xilinx usb host EHCI controller core"
 	depends on (PPC32 || MICROBLAZE)
