--- a/drivers/tty/serial/omap-serial.c	2016-07-12 18:00:13.000000000 +0300
+++ a/drivers/tty/serial/omap-serial.c	2016-07-14 14:18:42.048905600 +0300
@@ -494,6 +494,7 @@
 		if (status & UART_MSR_DDCD)
 			uart_handle_dcd_change
 				(&up->port, status & UART_MSR_DCD);
+dev_warn(up->dev, "%s(0): dcts: %d\n", __FUNCTION__, ( status & UART_MSR_DCTS));
 		if (status & UART_MSR_DCTS)
 			uart_handle_cts_change
 				(&up->port, status & UART_MSR_CTS);
@@ -1061,7 +1062,7 @@
 //dev_warn(up->dev, "%s(0): autorts\n", __FUNCTION__);
 		/* Enable AUTOCTS (autoRTS is enabled when RTS is raised) */
 		up->port.status |= UPSTAT_AUTOCTS | UPSTAT_AUTORTS;
-		up->efr |= UART_EFR_CTS;
+		if ( !( up->port.rs485.flags & SER_RS485_ENABLED)) up->efr |= UART_EFR_CTS;
 	} else {
 //dev_warn(up->dev, "%s(0): no autorts\n", __FUNCTION__);
 		/* Disable AUTORTS and AUTOCTS */
