--- a/tools/isp/isp.c	2020-07-02 14:30:16.970911364 +0300
+++ a/tools/isp/isp.c	2020-07-02 15:21:32.703088821 +0300
@@ -693,8 +693,8 @@
 				}
 
 				while (file_size) {
-					//size = (file_size > MAX_MEM_SIZE_FOR_ISP) ? MAX_MEM_SIZE_FOR_ISP : file_size;
-					size = (file_size > 0x4000000UL) ? 0x4000000UL : file_size;
+					size = (file_size > MAX_MEM_SIZE_FOR_ISP) ? MAX_MEM_SIZE_FOR_ISP : file_size;
+					//size = (file_size > 0x4000000UL) ? 0x4000000UL : file_size;
 					if (rootfs_gt_4GB && ((partition_programmed + size) > 0xfc000000UL)) {
 						size = 0xfc000000UL - partition_programmed;
 					}
