--- linux/arch/arm/boot/dts/am335x-TPPs3-t-sound.dts	1970-01-01 03:00:00.000000000 +0300
+++ linux/arch/arm/boot/dts/am335x-TPPs3-t-sound.dts	2014-12-16 10:12:59.961528417 +0300
@@ -0,0 +1,89 @@
+/*
+ * Tibbo Project PCB 3 for Linux (TPP3)
+ * Copyright (C) 2016 Tibbo Tech. - http://tibbo.com/tps/
+ * License: GPLv2 or later
+ */
+/dts-v1/;
+
+#include "am33xx.tpp.dtsi"
+#include "am335x-TPP.dtsi"
+
+/ {
+	model = "AM335x TPPs3 SOUND";
+	compatible = "ti,am335x-tpp3", "ti,am335x-tpp", "ti,am33xx";
+
+	sound {
+		compatible = "ti,da830-evm-audio";
+		ti,model = "3104 on AM335x TPP";
+		ti,audio-codec = <&tlv320aic3x>;
+		ti,mcasp-controller = <&mcasp0>;
+		ti,codec-clock-rate = /*<12000000>*/<24000000>;
+		ti,audio-routing =
+			"Headphone Jack",       "HPLOUT",
+			"Headphone Jack",       "HPROUT",
+/*			"MIC2L",                "Mic Bias",*/
+/*			"MIC2R",                "Mic Bias";*/
+			"MIC2L",                "Mic Jack",
+			"MIC2R",                "Mic Jack";
+	};
+
+	i2c-s11 {
+		pinctrl-names = "default";
+		pinctrl-0 = <&i2c_s11_pins>;
+		gpios = <&gpio1 23 GPIO_ACTIVE_HIGH	/* i2c_soft1.sda, gpio1_23, 0x5c (PIN_OUTPUT_PULLUP | MUX_MODE7) */
+			 &gpio2 26 GPIO_ACTIVE_HIGH	/* i2c_soft1.scl, gpio2_26, 0xf0 (PIN_INPUT | MUX_MODE7), old:gpio2 26|gpio1 28? GPIO_ACTIVE_HIGH */
+			>;
+		status = "okay";
+
+		tlv320aic3x: tlv320aic3x@18 {
+			compatible = "ti,tlv320aic3104";
+			reg = <0x18>;
+			status = "okay";
+			ai3x-micbias-vg = <2>;
+			gpio-reset = <&gpio0 10 GPIO_ACTIVE_HIGH>;
+
+			/* Regulators */
+			AVDD-supply = <&vaux2_reg>;
+			IOVDD-supply = <&vaux2_reg>;
+			DRVDD-supply = <&vio_reg>;
+			DVDD-supply = <&vio_reg>;
+		};
+	};
+
+};
+
+&am33xx_pinmux {
+	sound_pins: sound_pins {
+		pinctrl-single,pins = <
+			0xdc  (PIN_INPUT  | MUX_MODE3)	/* mcasp0_ahclkx   S9C */
+			0xc4  (PIN_INPUT  | MUX_MODE3)	/* mcasp0_fsx   S9B */
+			0xc0  (PIN_INPUT  | MUX_MODE3)	/* mcasp_aclkx ? S9A */
+			0x1a0 (PIN_OUTPUT  | MUX_MODE2)	/* mcasp_axr2  S11D */
+			0x68  (PIN_INPUT | MUX_MODE6)	/* mcasp_axr0  S11C */
+		>;
+	};
+};
+
+&mcasp0 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&sound_pins>;
+	status = "okay";
+
+	op-mode = <0>;          /* MCASP_IIS_MODE */
+	tdm-slots = <2>;
+	/* 4 serializers */
+	serial-dir = <  /* 0: INACTIVE, 1: TX, 2: RX */
+		2 0 1 0
+	>;
+	tx-num-evt = <32>;
+	rx-num-evt = <32>;
+};
+
+/* disabled: speed, free S29 */
+&usb {
+	status = "disabled";
+	control@44e10620 {  status = "disabled";  };
+	usb-phy@47401300 {  status = "disabled";  };
+	usb@47401000 {  status = "disabled";  };
+	dma-controller@47402000  {  status = "disabled";  };
+};
+