--- linux/arch/arm/boot/dts/am335x-TPPs3-t-usb-host.dts	1970-01-01 03:00:00.000000000 +0300
+++ linux/arch/arm/boot/dts/am335x-TPPs3-t-usb-host.dts	2014-12-16 10:12:59.961528417 +0300
@@ -0,0 +1,23 @@
+/*
+ * Tibbo Project PCB 3 for Linux (TPP3)
+ * Copyright (C) 2016 Tibbo Tech. - http://tibbo.com/tps/
+ * License: GPLv2 or later
+ */
+/dts-v1/;
+
+#include "am33xx.tpp.dtsi"
+#include "am335x-TPP.dtsi"
+
+/ {
+	model = "AM335x TPPs3 USB";
+	compatible = "ti,am335x-tpp3", "ti,am335x-tpp", "ti,am33xx";
+};
+
+/* disabled: speed, free S29 */
+&usb {
+	status = "okay";
+	control@44e10620 {  status = "okay";  };
+	usb-phy@47401300 {  status = "okay";  };
+	usb@47401000 {  dr_mode = "host";  status = "okay";  };
+	dma-controller@47402000  {  status = "okay";  };
+};
+