--- linux/arch/arm/boot/dts/am335x-TPPs3-USTC.dts	1970-01-01 03:00:00.000000000 +0300
+++ linux/arch/arm/boot/dts/am335x-TPPs3-USTC.dts	2014-12-16 10:12:59.961528417 +0300
@@ -0,0 +1,36 @@
+/*
+ * Tibbo Project PCB 3 for Linux (TPP3) +5 lines
+ * predefined configuration for Union Station Technology Center
+ * http://www.ustechcenter.com/
+ * Copyright (C) 2016 Tibbo Tech. - http://tibbo.com/tps/
+ * License: GPLv2 or later
+ */
+/dts-v1/;
+
+#include "am33xx.tpp.dtsi"
+#include "am335x-TPP.dtsi"
+
+/ {
+	model = "AM335x TPPs3 USTC";
+	compatible = "ti,am335x-tpp3", "ti,am335x-tpp", "ti,am33xx";
+
+	i2c-s05 {
+		pinctrl-names = "default";
+		pinctrl-0 = <&i2c_s05_pins>;
+		gpios = <&gpio2 28 GPIO_ACTIVE_HIGH	/* i2c_soft2.sda, gpio2_28, S5B, 0xf8 (PIN_OUTPUT_PULLUP | MUX_MODE7) */
+			 &gpio2 29 GPIO_ACTIVE_HIGH	/* i2c_soft2.scl, gpio2_29, S5A, 0xfc (PIN_INPUT | MUX_MODE7) */
+			>;
+	};
+	i2c-s07 {
+		pinctrl-names = "default";
+		pinctrl-0 = <&i2c_s07_pins>;
+		gpios = <&gpio0 20 GPIO_ACTIVE_HIGH	/* i2c_soft3.sda, gpio0_20, S7B, 0x1b4 (PIN_OUTPUT_PULLUP | MUX_MODE7) */
+			 &gpio2 27 GPIO_ACTIVE_HIGH	/* i2c_soft3.scl, gpio2_27, S7A, 0xf4 (PIN_INPUT | MUX_MODE7) */
+			>;
+	};
+
+};
+
+&uart1 {  status = "okay";  };  /* S1 (+S3) */
+&i2c_s05 {  status = "okay";  };
+&i2c_s07 {  status = "okay";  };
+