--- a/drivers/net/wireless/Kconfig	2019-08-07 16:41:14.234277723 +0300
+++ a/drivers/net/wireless/Kconfig	2019-08-07 16:45:04.970227445 +0300
@@ -45,6 +45,7 @@
 source "drivers/net/wireless/st/Kconfig"
 source "drivers/net/wireless/ti/Kconfig"
 source "drivers/net/wireless/zydas/Kconfig"
+source "drivers/net/wireless/bcmdhd/Kconfig"
 
 config PCMCIA_RAYCS
 	tristate "Aviator/Raytheon 2.4GHz wireless support"
