--- a/arch/arm/boot/dts/sp7021-common.dtsi	2020-08-26 11:53:16.716858111 +0300
+++ a/arch/arm/boot/dts/sp7021-common.dtsi	2020-08-26 14:50:46.665472570 +0300
@@ -10,10 +10,10 @@
 #include <dt-bindings/gpio/gpio.h>
 #include <dt-bindings/pinctrl/sppctl-sp7021.h>
 
-#define ENABLE_DMARX0
-#define ENABLE_DMARX1
-#define ENABLE_DMATX0
-#define ENABLE_DMATX1
+#define ENABLE_DMARX0
+#define ENABLE_DMARX1
+//#define ENABLE_DMATX0
+//#define ENABLE_DMATX1
 
 / {
 	aliases {
