--- a/drivers/spi/spi-sunplus-sp7021.c	2020-09-18 14:36:15.501368285 +0300
+++ a/drivers/spi/spi-sunplus-sp7021.c	2020-09-18 18:07:50.099990002 +0300
@@ -1436,8 +1436,9 @@
 
 	  
 	   if(spi->mode & SPI_CS_HIGH){
+DBG_INFO( "cs_high: 1\n");
 		  reg_temp = reg_temp | CS_POR;  
-	   }
+	   } else DBG_INFO( "cs_high: 0\n");
 	   if(spi->mode & SPI_LSB_FIRST){
 		  reg_temp = reg_temp | LSB_SEL;  
 	   }
