--- a/drivers/spi/spi-sunplus-sp7021.c	2020-09-20 19:08:51.107632178 +0300
+++ a/drivers/spi/spi-sunplus-sp7021.c	2020-09-21 20:30:53.192104676 +0300
@@ -1435,10 +1435,11 @@
 	   }
 
 	  
-	   if(!(spi->mode & SPI_CS_HIGH)){
-DBG_INFO( "cs_high: 1\n");
-		  reg_temp = reg_temp | CS_POR;  
-	   } else DBG_INFO( "cs_high: 0\n");
+//	   if(!(spi->mode & SPI_CS_HIGH)){
+//DBG_INFO( "cs_high: 1\n");
+//		  reg_temp = reg_temp | CS_POR;  
+//	   } else DBG_INFO( "cs_high: 0\n");
+reg_temp &= ~(1<<5);
 	   if(spi->mode & SPI_LSB_FIRST){
 		  reg_temp = reg_temp | LSB_SEL;  
 	   }
