--- a/Makefile	2022-04-07 12:56:07.726626947 +0300
+++ a/Makefile	2022-04-07 13:03:18.427359625 +0300
@@ -122,9 +122,9 @@
 install:
 	mkdir -p $(DESTDIR)$(libdir)/
 	if [ "$(PJ_EXCLUDE_PJSUA2)x" = "x" ] ; then \
-	    cp -af $(APP_LIBXX_FILES) $(DESTDIR)$(libdir)/; \
+	    install $(APP_LIBXX_FILES) $(DESTDIR)$(libdir)/; \
 	else \
-	    cp -af $(APP_LIB_FILES) $(DESTDIR)$(libdir)/; \
+	    install $(APP_LIB_FILES) $(DESTDIR)$(libdir)/; \
 	fi
 	mkdir -p $(DESTDIR)$(includedir)/
 	for d in pjlib pjlib-util pjnath pjmedia pjsip; do \
