--- a/drivers/net/wireless/libertas/main.c	2016-04-28 18:46:32.000000000 +0300
+++ a/drivers/net/wireless/libertas/main.c	2016-04-28 19:05:53.457305610 +0300
@@ -667,25 +667,34 @@
 	lbs_deb_enter(LBS_DEB_FW);
 
 	/* Read MAC address from firmware */
+pr_err("%s 0\n", __FUNCTION__);
 	eth_broadcast_addr(priv->current_addr);
+pr_err("%s 1\n", __FUNCTION__);
 	ret = lbs_update_hw_spec(priv);
+pr_err("%s 2:%d\n", __FUNCTION__, ret);
 	if (ret)
 		goto done;
 
+pr_err("%s 3\n", __FUNCTION__);
 	/* Read power levels if available */
 	ret = lbs_get_tx_power(priv, &curlevel, &minlevel, &maxlevel);
 	if (ret == 0) {
+pr_err("%s 4 0:%d 1:%d 2:%d\n", __FUNCTION__, curlevel, minlevel, maxlevel);
 		priv->txpower_cur = curlevel;
 		priv->txpower_min = minlevel;
 		priv->txpower_max = maxlevel;
 	}
 
+pr_err("%s 5\n", __FUNCTION__);
 	/* Send cmd to FW to enable 11D function */
 	ret = lbs_set_snmp_mib(priv, SNMP_MIB_OID_11D_ENABLE, 1);
+pr_err("%s 5:%d\n", __FUNCTION__, ret);
 	if (ret)
 		goto done;
 
+pr_err("%s 6\n", __FUNCTION__);
 	ret = lbs_set_mac_control_sync(priv);
+pr_err("%s 7:%d\n", __FUNCTION__, ret);
 done:
 	lbs_deb_leave_args(LBS_DEB_FW, "ret %d", ret);
 	return ret;
