--- a/drivers/net/wireless/libertas/main.c	2016-04-28 17:55:07.000000000 +0300
+++ a/drivers/net/wireless/libertas/main.c	2016-04-28 18:43:33.220172076 +0300
@@ -1104,25 +1104,32 @@
 	lbs_deb_enter(LBS_DEB_MAIN);
 
 	/* poke the firmware */
+pr_err("%s 0\n", __FUNCTION__);
 	ret = lbs_setup_firmware(priv);
 	if (ret)
 		goto done;
 
-	if (!lbs_disablemesh)
+pr_err("%s 1\n", __FUNCTION__);
+	if (!lbs_disablemesh) {
+pr_err("%s 1-0\n", __FUNCTION__);
 		lbs_init_mesh(priv);
-	else
+	} else
 		pr_info("%s: mesh disabled\n", dev->name);
 
+pr_err("%s 2\n", __FUNCTION__);
 	if (lbs_cfg_register(priv)) {
 		pr_err("cannot register device\n");
 		goto done;
 	}
 
+pr_err("%s 3\n", __FUNCTION__);
 	if (lbs_mesh_activated(priv))
 		lbs_start_mesh(priv);
 
+pr_err("%s 4\n", __FUNCTION__);
 	lbs_debugfs_init_one(priv, dev);
 
+pr_err("%s 5\n", __FUNCTION__);
 	netdev_info(dev, "Marvell WLAN 802.11 adapter\n");
 
 	ret = 0;
