--- a/drivers/usb/core/devices.c	2017-09-20 09:25:38.000000000 +0300
+++ a/drivers/usb/core/devices.c	2018-11-12 06:42:55.000000000 +0300
@@ -58,6 +58,7 @@
 #include <linux/usb/hcd.h>
 #include <linux/mutex.h>
 #include <linux/uaccess.h>
+#include <linux/usb/sp_usb.h>
 
 #include "usb.h"
 
@@ -496,7 +497,7 @@
 	if (*nbytes <= 0)
 		return 0;
 
-	if (level > MAX_TOPO_LEVEL)
+	if (level > max_topo_level)
 		return 0;
 	/* allocate 2^1 pages = 8K (on i386);
 	 * should be more than enough for one device */
