--- a/drivers/usb/core/hub.h	2017-09-20 09:25:38.000000000 +0300
+++ a/drivers/usb/core/hub.h	2018-11-12 06:42:55.000000000 +0300
@@ -41,6 +41,7 @@
 	int			error;		/* last reported error */
 	int			nerrors;	/* track consecutive errors */
 
+	struct list_head	event_list;	/* hubs w/data or errs ready */
 	unsigned long		event_bits[1];	/* status change bitmask */
 	unsigned long		change_bits[1];	/* ports with logical connect
 							status change */
@@ -78,6 +79,9 @@
 	struct delayed_work	init_work;
 	struct work_struct      events;
 	struct usb_port		**ports;
+#ifdef	CONFIG_USB_LOGO_TEST
+	struct task_struct	*usb_logo_thread;
+#endif
 };
 
 /**
