--- a/drivers/usb/core/Kconfig	2017-09-20 09:25:38.000000000 +0300
+++ a/drivers/usb/core/Kconfig	2018-11-12 06:42:55.000000000 +0300
@@ -1,6 +1,23 @@
 #
 # USB Core configuration
 #
+config USB_DEBUG
+	bool "USB verbose debug messages"
+	depends on USB
+	help
+	  Say Y here if you want the USB core & hub drivers to produce a bunch
+	  of debug messages to the system log. Select this if you are having a
+	  problem with USB support and want to see more of what is going on.
+
+config USB_PHY_RX_ACTIVE_QUESTION_WORKAROUND
+	bool "Enable USB PHY RX Active Question Workaround"
+	depends on USB
+
+config USB_BAD_DEVICE_INFO
+	bool "support bad usb device info"
+	depends on USB
+	default n
+
 config USB_ANNOUNCE_NEW_DEVICES
 	bool "USB announce new devices"
 	help
@@ -43,7 +60,6 @@
 
 config USB_OTG
 	bool "OTG support"
-	depends on PM
 	default n
 	help
 	  The most notable feature of USB OTG is support for a
