--- linux/arch/arm/boot/dts/am335x-TPPs3-RLP.dts	1970-01-01 03:00:00.000000000 +0300
+++ linux/arch/arm/boot/dts/am335x-TPPs3-RLP.dts	2014-12-16 10:12:59.961528417 +0300
@@ -0,0 +1,66 @@
+/*
+ * TPP3 default 1-uart+gpio configuration
+ * Copyright (C) 2016 Tibbo Tech. - http://tibbo.com/tps/
+ * License: GPLv2 or later
+ */
+/dts-v1/;
+
+#include "am33xx.tpp.dtsi"
+#include "am335x-TPP.dtsi"
+#include "am335x-TPPs3.dtsi"
+
+/ {
+	model = "AM335x TPPs3 RusElProm";
+	compatible = "ti,am335x-tpp3", "ti,am335x-tpp", "ti,am33xx";
+};
+
+&am33xx_pinmux {
+	pinctrl-0 = <&gpio_pins>;
+	gpio_pins: gpio_pins {
+		pinctrl-single,pins = <
+			/* 0x78  (PIN_OUTPUT | INPUT_EN | MUX_MODE7) */
+			/* 0x8c  (PIN_OUTPUT | INPUT_EN | MUX_MODE7) */
+			/* 0x234 (PIN_OUTPUT | INPUT_EN | MUX_MODE7) */
+			AM33XX_IOPAD(0xa34, (PIN_INPUT_PULLUP | INPUT_EN | MUX_MODE7))	/* gpio3.13 pin109 S3D (usb1) */
+		>;
+	};
+	sss_pins: sss_pins {
+		pinctrl-single,pins = <
+			0x164 (PIN_INPUT_PULLUP | MUX_MODE4) /* g 0.07 */
+			0x168 (PIN_INPUT_PULLDOWN | MUX_MODE4) /* g 1.08 */
+			0x16c (PIN_INPUT_PULLUP | MUX_MODE4) /* g 1.09 */
+			0x144 (PIN_INPUT_PULLDOWN | MUX_MODE2) /* g 0.29 */
+		>;
+	};
+};
+
+&uart0 {  status = "okay";  };
+/*&spi0 {  status = "okay";  };*/
+/* hw spi - for test */
+/*
+&spi1 {
+ status = "disabled";
+ pinctrl-names = "default";
+ pinctrl-0 = <&sss_pins>;
+ ade7758@0 {
+      compatible = "adi,ade7758";
+      reg = <0>;
+      spi-max-frequency = <500000>;
+ };
+};
+*/
+&i2c_s07 {  status = "okay";  };
+&mmc3 {  status = "okay";  };
+&spi_s19 {
+ status = "okay";
+ spi@0 {  status = "disabled";  };
+ ade7758@0 {
+      compatible = "adi,ade7758"/*"adi,ade7854"*/;
+      reg = <0>;
+      spi-max-frequency = <500000>;
+ };
+};
+/* */
+&i2c_s13 {  status = "okay";  };
+&i2c_s09 {  status = "okay";  };
+&i2c_s11 {  status = "okay";  };
+