--- linux/arch/arm/boot/dts/am335x-TPP.dtsi	1970-01-01 03:00:00.000000000 +0300
+++ linux/arch/arm/boot/dts/am335x-TPP.dtsi	2014-12-17 01:12:53.888613349 +0300
@@ -0,0 +1,674 @@
+/*
+ * Tibbo Project PCB for Linux (TPP) base definitions + 4 lines
+ * Copyright (C) 2016 Tibbo Tech. - http://tibbo.com/tps/
+ * License: GPLv2 or later
+ * Based on am33xx-evm.dtsi for am335x SoC
+ */
+
+#include <dt-bindings/interrupt-controller/irq.h>
+
+/ {
+	model = "AM335x TPP";
+	compatible = "ti,am335x-tpp", "ti,am33xx";
+
+	cpus {
+		cpu@0 {
+			cpu0-supply = <&vdd1_reg>;
+		};
+	};
+
+	memory {
+		device_type = "memory";
+		reg = <0x80000000 0x20000000>; /* 512 MB */
+	};
+
+	chosen {
+		stdout-path = &uart0;
+	};
+
+	leds {
+		pinctrl-names = "default";
+		pinctrl-0 = <&user_leds_s0>;
+		compatible = "gpio-leds";
+		led@2 {
+			label = "SG";
+			gpios = <&gpio1 24 GPIO_ACTIVE_HIGH>;
+			linux,default-trigger = "heartbeat";
+			default-state = "off";
+		};
+		led@3 {
+			label = "SR";
+			gpios = <&gpio1 22 GPIO_ACTIVE_LOW>;
+			linux,default-trigger = "mmc0";
+			default-state = "off";
+		};
+
+	};
+
+	vbat: fixedregulator@0 {
+		compatible = "regulator-fixed";
+		regulator-name = "vbat";
+		regulator-min-microvolt = <5000000>;
+		regulator-max-microvolt = <5000000>;
+		regulator-boot-on;
+	};
+
+	buzzer0 {
+		compatible = "pwm-beeper";
+		pinctrl-names = "default";
+		pinctrl-0 = <&buzzer0_pins>;
+		pwms = <&ehrpwm0 1 1000000 0>;
+	};
+	tpsleds: tpsleds {
+		compatible = "tpsleds";
+		pinctrl-names = "default";
+		pinctrl-0 = <&tpsleds_pins>;
+		status = "okay";
+		init-val = <0x0>;
+		gpios = <&gpio1 27 GPIO_ACTIVE_HIGH
+			 &gpio1 25 GPIO_ACTIVE_HIGH
+		>;
+	};
+
+	i2c_tps: i2c-tps { /* index:i2c-3 | tps, eeprom */
+		compatible = "i2c-gpio";
+/*		i2c-gpio,scl-open-drain;*/
+/*		i2c-gpio,sda-open-drain;*/
+		i2c-gpio,delay-us = <2>;	/* ~100 kHz */
+		#address-cells = <1>;
+		#size-cells = <0>;
+		status = "disabled";
+	};
+
+	tps_wifi: tps-wifi {
+		compatible = "spi-gpio";
+		#address-cells = <0x1>;
+		#size-cells = <0x0>;
+		/*ranges;*/
+		cs-gpios  = <&gpio0 6 GPIO_ACTIVE_HIGH>; /* wifi_cs   gpio0.06 0x160 */
+		gpio-sck  = <&gpio0 2 GPIO_ACTIVE_HIGH>; /* wifi_clk  gpio0.02 0x150 */
+		gpio-mosi = <&gpio0 4 GPIO_ACTIVE_HIGH>; /* wifi_mosi gpio0.04 0x158 */
+		gpio-miso = <&gpio0 3 GPIO_ACTIVE_HIGH>; /* wifi_miso gpio0.03 0x154 */
+		num-chipselects = <1>;
+		pinctrl-names = "default";
+		pinctrl-0 = <&spi_wifi_pins &spi_wifi_add_gpio>;
+		status = /*"okay"*/"disabled";
+	};
+	spi_s05: spi-s05 {
+		compatible = "spi-gpio";
+		#address-cells = <0x1>;
+		#size-cells = <0x0>;
+		/*ranges;*/
+		cs-gpios  = <&gpio2 29 GPIO_ACTIVE_LOW>;  /* S05A gpio2.29 0xfc */
+		gpio-sck  = <&gpio2 28 GPIO_ACTIVE_HIGH>; /* S05B gpio2.28 0xf8 */
+		gpio-mosi = <&gpio2 17 GPIO_ACTIVE_HIGH>; /* S05C gpio2.17 0xcc */
+		gpio-miso = <&gpio2 16 GPIO_ACTIVE_HIGH>; /* S05D gpio2.16 0xc8 */
+		num-chipselects = <1>;
+		pinctrl-names = "default";
+		pinctrl-0 = <&spi_s05_pins>;
+		status = "disabled";
+		spi@0 {
+			compatible = /*"spidev"*/"rohm,dh2228fv";
+			reg = <0>;
+			spi-max-frequency = <57600000>;
+		};
+	};
+
+};
+
+&am33xx_pinmux {
+	pinctrl-names = "default";
+/*	pinctrl-0 = <&gpio_pins &clkout2_pin>;*/
+	pinctrl-0 = <&gpio_pins &timer_pins/* &timer6_pins*/>;
+
+	tpsleds_pins: tpsleds_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x86c, (PIN_OUTPUT | MUX_MODE7))	/* blue led data  gpio1_27 */
+			AM33XX_IOPAD(0x864, (PIN_OUTPUT | MUX_MODE7))	/* blue led clk   gpio1_25 */
+		>;
+	};
+	timer_pins: timer_pins {
+		pinctrl-single,pins = <
+/*			AM33XX_IOPAD(0x9b0, (PIN_OUTPUT_PULLDOWN | MUX_MODE2))*/	/* gpio0.19 pin19  S3A  timer4 mux1 */
+/* remove later: power for USB */
+/*			AM33XX_IOPAD(0xa1c, (PIN_OUTPUT | MUX_MODE0))*/	/* gpio0.18, pin19  S3D */
+/* remove later: pwm */
+/*			AM33XX_IOPAD(0x954, (PIN_OUTPUT | MUX_MODE3))*/	/* gpio0.03, wireless topright pin */
+/* remove later: button */
+/*			AM33XX_IOPAD(0x8a8, (PIN_INPUT | MUX_MODE7))*/	/* 'program' button */
+		>;
+	};
+	timer4_pins: timer4_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x9b0, (PIN_OUTPUT | MUX_MODE2))	/* gpio0.19 pin19  S3A  timer4 mux1 */
+		>;
+	};
+	timer5_pins: timer5_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x8f0, (PIN_OUTPUT_PULLUP | MUX_MODE3))	/* gpio2.26 pin90  S11A timer5 mux1 */
+		>;
+	};
+	timer6_pins: timer6_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x8f4, (PIN_OUTPUT_PULLDOWN | MUX_MODE3))	/* gpio2.27 pin91  S7A  timer6 mux1 */
+		>;
+	};
+	timer7_pins: timer7_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x9b4, (PIN_OUTPUT_PULLDOWN | MUX_MODE4))	/* gpio0.20 pin20  S7B  timer7 mux1 */
+		>;
+	};
+	i2c_tps_pins: i2c_tps_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x840, (PIN_INPUT | MUX_MODE7))	/* has to be 3f|2f no S* */
+			AM33XX_IOPAD(0x854, (PIN_INPUT | MUX_MODE7))	/* has to be 3f|2f no S* */
+		>;
+	};
+	spi_wifi_add_gpio: spi_wifi_add_gpio {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x1a8, (PIN_OUTPUT | MUX_MODE7))	/* gpio3.20 pin116 wifi_reset */
+			AM33XX_IOPAD(0x1a4, (PIN_INPUT | MUX_MODE7))	/* gpio3.19 pin115 wifi_irq (extra1) */
+		>;
+	};
+	spi_wifi_pins: spi_wifi_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x960, (PIN_OUTPUT | MUX_MODE7))	/* gpio0.06 pin003 wifi_cs */
+			AM33XX_IOPAD(0x950, (PIN_OUTPUT | MUX_MODE7))	/* gpio0.02 pin010 wifi_clk */
+			AM33XX_IOPAD(0x958, (PIN_INPUT  | MUX_MODE7))	/* gpio0.04 pin005 wifi_mosi */
+			AM33XX_IOPAD(0x954, (PIN_OUTPUT | MUX_MODE7))	/* gpio0.03 pin009 wifi_miso */
+		>;
+	};
+	spi_wifi_pinsS: spi_wifi_pinsS {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x960, (PIN_INPUT_PULLUP   | MUX_MODE0))	/* gpio0.06 pin003 wifi_cs */
+			AM33XX_IOPAD(0x950, (PIN_INPUT_PULLDOWN | MUX_MODE0))	/* gpio0.02 pin010 wifi_clk */
+			AM33XX_IOPAD(0x958, (PIN_INPUT_PULLUP   | MUX_MODE0))	/* gpio0.04 pin005 wifi_mosi */
+			AM33XX_IOPAD(0x954, (PIN_INPUT_PULLDOWN | MUX_MODE0))	/* gpio0.03 pin009 wifi_miso */
+		>;
+	};
+	spi_s05_pins: spi_s05_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x8fc, (PIN_INPUT_PULLUP   | MUX_MODE7))	/* gpio2.29 pin093 S5A */
+			AM33XX_IOPAD(0x8f8, (PIN_INPUT_PULLDOWN | MUX_MODE7))	/* gpio2.28 pin092 S5B */
+			AM33XX_IOPAD(0x8cc, (PIN_INPUT_PULLUP   | MUX_MODE7))	/* gpio2.17 pin081 S5C */
+			AM33XX_IOPAD(0x8c8, (PIN_INPUT_PULLDOWN | MUX_MODE7))	/* gpio2.16 pin080 S5D */
+		>;
+	};
+	gpio_pins: gpio_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x880,  (PIN_OUTPUT | MUX_MODE7))	/* gpio1.30 pin62  S19D */
+			AM33XX_IOPAD(0x884, (PIN_OUTPUT | MUX_MODE7))	/* gpio1.31 pin63  S19C */
+			/*AM33XX_IOPAD(0x8f4,  (PIN_OUTPUT | MUX_MODE7))*/	/* gpio2.27 pin91  S11A */
+			/*AM33XX_IOPAD(0x9b4, (PIN_OUTPUT | MUX_MODE7))*/	/* gpio0.20 pin20  S11B or see clkout2 below */
+			AM33XX_IOPAD(0x95c, (PIN_OUTPUT | MUX_MODE7))	/* gpio0.05 pin5   S21D */
+			/*AM33XX_IOPAD(0xa34, (PIN_OUTPUT | MUX_MODE7))*/	/* gpio3.13 pin109 S25A */
+			/*AM33XX_IOPAD(0x960, (PIN_OUTPUT | MUX_MODE7))*/	/* gpio0.06 pin06  W003 */
+			/*AM33XX_IOPAD(0x958, (PIN_OUTPUT | MUX_MODE7))*/	/* gpio0.04 pin04  W005 */
+			AM33XX_IOPAD(0x9a8, (PIN_OUTPUT | MUX_MODE7))	/* gpio3.20 pin116 W007 */
+			/*AM33XX_IOPAD(0x954, (PIN_OUTPUT | MUX_MODE7))*/	/* gpio0.03 pin03  W009 */
+			/*AM33XX_IOPAD(0x950, (PIN_OUTPUT | MUX_MODE7))*/	/* gpio0.02 pin02  W010 */
+			AM33XX_IOPAD(0x9a4, (PIN_OUTPUT | MUX_MODE7))	/* gpio3.19 pin115 W008 */
+/*			AM33XX_IOPAD(0xa1c, (PIN_OUTPUT | MUX_MODE7))*/	/* gpio0.18 pin??? S25A */
+		>;
+	};
+	buzzer0_pins: buzzer0_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x994, (PIN_OUTPUT | MUX_MODE1))	/* dedicated buzzer pin: gpio101 in /sys/../pinmux/pins */
+		>;
+	};
+	user_leds_s0: user_leds_s0 {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x858, (PIN_OUTPUT_PULLUP | MUX_MODE7))	/* gpmc_a6.gpio1_22 */
+			AM33XX_IOPAD(0x860, (PIN_OUTPUT_PULLUP | MUX_MODE7))	/* gpmc_a8.gpio1_24 */
+		>;
+	};
+
+	usb0_pins: usb0_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0xa1c, (PIN_OUTPUT | MUX_MODE0))	/* gpio0.18 pin??? S25A */
+		>;
+	};
+	usb1_pins: usb1_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0xa34, (PIN_OUTPUT | MUX_MODE0))	/* gpio0.18 pin??? S25A or S3D ??? */
+		>;
+	};
+
+/* nothing currently connected to hw I2C */
+	i2c0_pins: i2c0_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x988, (PIN_INPUT_PULLUP | MUX_MODE0))	/* i2c0_sda.i2c0_sda */
+			AM33XX_IOPAD(0x98c, (PIN_INPUT_PULLUP | MUX_MODE0))	/* i2c0_scl.i2c0_scl */
+		>;
+	};
+
+	clkout2_pin: pinmux_clkout2_pin {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x9b4, (PIN_OUTPUT_PULLDOWN | MUX_MODE3))	/* xdma_event_intr1.clkout2 */
+		>;
+	};
+
+	cpsw_default: cpsw_default {
+		pinctrl-single,pins = <
+			/* Slave 1 */
+			AM33XX_IOPAD(0x908, (PIN_INPUT_PULLUP | MUX_MODE7/*MUX_MODE0*/))	/* mii1_col.mii1_col */
+			AM33XX_IOPAD(0x90C, (PIN_INPUT_PULLUP | MUX_MODE7/*MUX_MODE0*/))	/* mii1_crs.mii1_crs */
+			AM33XX_IOPAD(0x910, (PIN_INPUT_PULLUP | MUX_MODE0))	/* mii1_rxerr.mii1_rxerr */
+			AM33XX_IOPAD(0x914, (PIN_OUTPUT_PULLDOWN | MUX_MODE0))	/* mii1_txen.mii1_txen */
+			AM33XX_IOPAD(0x918, (PIN_INPUT_PULLUP | MUX_MODE0))	/* mii1_rxdv.mii1_rxdv */
+			AM33XX_IOPAD(0x91c, (PIN_OUTPUT_PULLDOWN | MUX_MODE0))	/* mii1_txd3.mii1_txd3 */
+			AM33XX_IOPAD(0x920, (PIN_OUTPUT_PULLDOWN | MUX_MODE0))	/* mii1_txd2.mii1_txd2 */
+			AM33XX_IOPAD(0x924, (PIN_OUTPUT_PULLDOWN | MUX_MODE0))	/* mii1_txd1.mii1_txd1 */
+			AM33XX_IOPAD(0x928, (PIN_OUTPUT_PULLDOWN | MUX_MODE0))	/* mii1_txd0.mii1_txd0 */
+			AM33XX_IOPAD(0x92c, (PIN_INPUT_PULLUP | MUX_MODE0))	/* mii1_txclk.mii1_txclk */
+			AM33XX_IOPAD(0x930, (PIN_INPUT_PULLUP | MUX_MODE0))	/* mii1_rxclk.mii1_rxclk */
+			AM33XX_IOPAD(0x934, (PIN_INPUT_PULLUP | MUX_MODE0))	/* mii1_rxd3.mii1_rxd3 */
+			AM33XX_IOPAD(0x938, (PIN_INPUT_PULLUP | MUX_MODE0))	/* mii1_rxd2.mii1_rxd2 */
+			AM33XX_IOPAD(0x93c, (PIN_INPUT_PULLUP | MUX_MODE0))	/* mii1_rxd1.mii1_rxd1 */
+			AM33XX_IOPAD(0x940, (PIN_INPUT_PULLUP | MUX_MODE0))	/* mii1_rxd0.mii1_rxd0 */
+		>;
+	};
+
+	cpsw_sleep: cpsw_sleep {
+		pinctrl-single,pins = <
+			/* Slave 1 reset value */
+			AM33XX_IOPAD(0x908, (PIN_INPUT_PULLDOWN | MUX_MODE7))
+			AM33XX_IOPAD(0x90C, (PIN_INPUT_PULLDOWN | MUX_MODE7))
+			AM33XX_IOPAD(0x910, (PIN_INPUT_PULLDOWN | MUX_MODE7))
+			AM33XX_IOPAD(0x914, (PIN_INPUT_PULLDOWN | MUX_MODE7))
+			AM33XX_IOPAD(0x918, (PIN_INPUT_PULLDOWN | MUX_MODE7))
+			AM33XX_IOPAD(0x91c, (PIN_INPUT_PULLDOWN | MUX_MODE7))
+			AM33XX_IOPAD(0x920, (PIN_INPUT_PULLDOWN | MUX_MODE7))
+			AM33XX_IOPAD(0x924, (PIN_INPUT_PULLDOWN | MUX_MODE7))
+			AM33XX_IOPAD(0x928, (PIN_INPUT_PULLDOWN | MUX_MODE7))
+			AM33XX_IOPAD(0x92c, (PIN_INPUT_PULLDOWN | MUX_MODE7))
+			AM33XX_IOPAD(0x930, (PIN_INPUT_PULLDOWN | MUX_MODE7))
+			AM33XX_IOPAD(0x934, (PIN_INPUT_PULLDOWN | MUX_MODE7))
+			AM33XX_IOPAD(0x938, (PIN_INPUT_PULLDOWN | MUX_MODE7))
+			AM33XX_IOPAD(0x93c, (PIN_INPUT_PULLDOWN | MUX_MODE7))
+			AM33XX_IOPAD(0x940, (PIN_INPUT_PULLDOWN | MUX_MODE7))
+		>;
+	};
+
+	davinci_mdio_default: davinci_mdio_default {
+		pinctrl-single,pins = <
+			/* MDIO */
+			AM33XX_IOPAD(0x948, (PIN_INPUT_PULLUP | SLEWCTRL_FAST | MUX_MODE0))	/* mdio_data.mdio_data */
+			AM33XX_IOPAD(0x94c, (PIN_OUTPUT_PULLUP | MUX_MODE0))			/* mdio_clk.mdio_clk */
+		>;
+	};
+	davinci_mdio_sleep: davinci_mdio_sleep {
+		pinctrl-single,pins = <
+			/* MDIO reset value */
+			AM33XX_IOPAD(0x948, (PIN_INPUT_PULLDOWN | MUX_MODE7))
+			AM33XX_IOPAD(0x94c, (PIN_INPUT_PULLDOWN | MUX_MODE7))
+		>;
+	};
+
+	nandflash_pins_s0: nandflash_pins_s0 {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x800, (PIN_INPUT_PULLUP | MUX_MODE0))	/* gpmc_ad0.gpmc_ad0 */
+			AM33XX_IOPAD(0x804, (PIN_INPUT_PULLUP | MUX_MODE0))	/* gpmc_ad1.gpmc_ad1 */
+			AM33XX_IOPAD(0x808, (PIN_INPUT_PULLUP | MUX_MODE0))	/* gpmc_ad2.gpmc_ad2 */
+			AM33XX_IOPAD(0x80c, (PIN_INPUT_PULLUP | MUX_MODE0))	/* gpmc_ad3.gpmc_ad3 */
+			AM33XX_IOPAD(0x810, (PIN_INPUT_PULLUP | MUX_MODE0))	/* gpmc_ad4.gpmc_ad4 */
+			AM33XX_IOPAD(0x814, (PIN_INPUT_PULLUP | MUX_MODE0))	/* gpmc_ad5.gpmc_ad5 */
+			AM33XX_IOPAD(0x818, (PIN_INPUT_PULLUP | MUX_MODE0))	/* gpmc_ad6.gpmc_ad6 */
+			AM33XX_IOPAD(0x81c, (PIN_INPUT_PULLUP | MUX_MODE0))	/* gpmc_ad7.gpmc_ad7 */
+			AM33XX_IOPAD(0x820, (PIN_INPUT_PULLUP | MUX_MODE0))	/* gpmc_ad8.gpmc_ad8 */
+			AM33XX_IOPAD(0x824, (PIN_INPUT_PULLUP | MUX_MODE0))	/* gpmc_ad9.gpmc_ad9 */
+			AM33XX_IOPAD(0x828, (PIN_INPUT_PULLUP | MUX_MODE0))	/* gpmc_ad10.gpmc_ad10 */
+			AM33XX_IOPAD(0x82c, (PIN_INPUT_PULLUP | MUX_MODE0))	/* gpmc_ad11.gpmc_ad11 */
+			AM33XX_IOPAD(0x830, (PIN_INPUT_PULLUP | MUX_MODE0))	/* gpmc_ad12.gpmc_ad12 */
+			AM33XX_IOPAD(0x834, (PIN_INPUT_PULLUP | MUX_MODE0))	/* gpmc_ad13.gpmc_ad13 */
+			AM33XX_IOPAD(0x838, (PIN_INPUT_PULLUP | MUX_MODE0))	/* gpmc_ad14.gpmc_ad14 */
+			AM33XX_IOPAD(0x83c, (PIN_INPUT_PULLUP | MUX_MODE0))	/* gpmc_ad15.gpmc_ad15 */
+			AM33XX_IOPAD(0x870, (PIN_INPUT_PULLUP | MUX_MODE0))	/* gpmc_wait0.gpmc_wait0 */
+			/*AM33XX_IOPAD(0x874, (PIN_INPUT_PULLUP | MUX_MODE7))*/	/* gpmc_wpn.gpio0_30 - not works */
+			AM33XX_IOPAD(0x87c, (PIN_OUTPUT | MUX_MODE0))		/* gpmc_csn0.gpmc_csn0  */
+			AM33XX_IOPAD(0x890, (PIN_OUTPUT | MUX_MODE0))		/* gpmc_advn_ale.gpmc_advn_ale */
+			AM33XX_IOPAD(0x894, (PIN_OUTPUT | MUX_MODE0))		/* gpmc_oen_ren.gpmc_oen_ren */
+			AM33XX_IOPAD(0x898, (PIN_OUTPUT | MUX_MODE0))		/* gpmc_wen.gpmc_wen */
+			AM33XX_IOPAD(0x89c, (PIN_OUTPUT | MUX_MODE0))		/* gpmc_be0n_cle.gpmc_be0n_cle */
+		>;
+	};
+
+	mmc2_pins: mmc2_pins {	/* for mmc3 */
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x844, (PIN_INPUT_PULLUP | MUX_MODE3))	/* mmc2_d0  S23_A */
+			AM33XX_IOPAD(0x848, (PIN_INPUT_PULLUP | MUX_MODE3))	/* mmc2_d1  S23_B */
+			AM33XX_IOPAD(0x84c, (PIN_INPUT_PULLUP | MUX_MODE3))	/* mmc2_d2  S23_C */
+			AM33XX_IOPAD(0x878, (PIN_INPUT_PULLUP | MUX_MODE3))	/* mmc2_d3  S23_D */
+			AM33XX_IOPAD(0x888, (PIN_INPUT_PULLUP | MUX_MODE3))	/* mmc2_cmd S21_B */
+			AM33XX_IOPAD(0x88c, (PIN_INPUT_PULLUP | MUX_MODE3))	/* mmc2_clk S21_A */
+			AM33XX_IOPAD(0x998, (PIN_INPUT_PULLUP | MUX_MODE7))	/* mmc2_cd  S21_C */
+		>;
+	};
+	mmc2_idle: mmc2_idle {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x998, (PIN_INPUT_PULLUP | MUX_MODE7))	/* mmc2_cd  S21_C */
+		>;
+	};
+	hdmi_pins: hdmi_pins {
+		pinctrl-single,pins = <
+			/*AM33XX_IOPAD(0x9b0, (MUX_MODE3))*/		/* xdma_event_intr0, OMAP_MUX_MODE3 | AM33XX_PIN_OUTPUT */
+			AM33XX_IOPAD(0x8a0, (PIN_OUTPUT | MUX_MODE0))	/* lcd_data0.lcd_data0, OMAP_MUX_MODE0 | AM33XX_PIN_OUTPUT | AM33XX_PULL_DISA */
+			AM33XX_IOPAD(0x8a4, (PIN_OUTPUT | MUX_MODE0))	/* lcd_data1.lcd_data1, OMAP_MUX_MODE0 | AM33XX_PIN_OUTPUT | AM33XX_PULL_DISA */
+			AM33XX_IOPAD(0x8a8, (PIN_OUTPUT | MUX_MODE0))	/* lcd_data2.lcd_data2, OMAP_MUX_MODE0 | AM33XX_PIN_OUTPUT | AM33XX_PULL_DISA */
+			AM33XX_IOPAD(0x8ac, (PIN_OUTPUT | MUX_MODE0))	/* lcd_data3.lcd_data3, OMAP_MUX_MODE0 | AM33XX_PIN_OUTPUT | AM33XX_PULL_DISA */
+			AM33XX_IOPAD(0x8b0, (PIN_OUTPUT | MUX_MODE0))	/* lcd_data4.lcd_data4, OMAP_MUX_MODE0 | AM33XX_PIN_OUTPUT | AM33XX_PULL_DISA */
+			AM33XX_IOPAD(0x8b4, (PIN_OUTPUT | MUX_MODE0))	/* lcd_data5.lcd_data5, OMAP_MUX_MODE0 | AM33XX_PIN_OUTPUT | AM33XX_PULL_DISA */
+			AM33XX_IOPAD(0x8b8, (PIN_OUTPUT | MUX_MODE0))	/* lcd_data6.lcd_data6, OMAP_MUX_MODE0 | AM33XX_PIN_OUTPUT | AM33XX_PULL_DISA */
+			AM33XX_IOPAD(0x8bc, (PIN_OUTPUT | MUX_MODE0))	/* lcd_data7.lcd_data7, OMAP_MUX_MODE0 | AM33XX_PIN_OUTPUT | AM33XX_PULL_DISA */
+			AM33XX_IOPAD(0x8c0, (PIN_OUTPUT | MUX_MODE0))	/* lcd_data8.lcd_data8, OMAP_MUX_MODE0 | AM33XX_PIN_OUTPUT | AM33XX_PULL_DISA */
+			AM33XX_IOPAD(0x8c4, (PIN_OUTPUT | MUX_MODE0))	/* lcd_data9.lcd_data9, OMAP_MUX_MODE0 | AM33XX_PIN_OUTPUT | AM33XX_PULL_DISA */
+			AM33XX_IOPAD(0x8c8, (PIN_OUTPUT | MUX_MODE0))	/* lcd_data10.lcd_data10, OMAP_MUX_MODE0 | AM33XX_PIN_OUTPUT | AM33XX_PULL_DISA */
+			AM33XX_IOPAD(0x8cc, (PIN_OUTPUT | MUX_MODE0))	/* lcd_data11.lcd_data11, OMAP_MUX_MODE0 | AM33XX_PIN_OUTPUT | AM33XX_PULL_DISA */
+			AM33XX_IOPAD(0x8d0, (PIN_OUTPUT | MUX_MODE0))	/* lcd_data12.lcd_data12, OMAP_MUX_MODE0 | AM33XX_PIN_OUTPUT | AM33XX_PULL_DISA */
+			AM33XX_IOPAD(0x8d4, (PIN_OUTPUT | MUX_MODE0))	/* lcd_data13.lcd_data13, OMAP_MUX_MODE0 | AM33XX_PIN_OUTPUT | AM33XX_PULL_DISA */
+			AM33XX_IOPAD(0x8d8, (PIN_OUTPUT | MUX_MODE0))	/* lcd_data14.lcd_data14, OMAP_MUX_MODE0 | AM33XX_PIN_OUTPUT | AM33XX_PULL_DISA */
+			AM33XX_IOPAD(0x8dc, (PIN_OUTPUT | MUX_MODE0))	/* lcd_data15.lcd_data15, OMAP_MUX_MODE0 | AM33XX_PIN_OUTPUT | AM33XX_PULL_DISA */
+			AM33XX_IOPAD(0x8e0, (PIN_OUTPUT | MUX_MODE0))	/* lcd_vsync.lcd_vsync, OMAP_MUX_MODE0 | AM33XX_PIN_OUTPUT */
+			AM33XX_IOPAD(0x8e4, (PIN_OUTPUT | MUX_MODE0))	/* lcd_hsync.lcd_hsync, OMAP_MUX_MODE0 | AM33XX_PIN_OUTPUT */
+			AM33XX_IOPAD(0x8e8, (PIN_OUTPUT | MUX_MODE0))	/* lcd_pclk.lcd_pclk, OMAP_MUX_MODE0 | AM33XX_PIN_OUTPUT */
+			AM33XX_IOPAD(0x8ec, (PIN_OUTPUT | MUX_MODE7/*0*/))	/* lcd_ac_bias_en.lcd_ac_bias_en, OMAP_MUX_MODE0 | AM33XX_PIN_OUTPUT */
+		>;
+	};
+
+};
+
+&usb {
+	usb-phy@47401300 {
+		pinctrl-names = "default";
+		pinctrl-0 = <&usb0_pins>;
+	};
+	usb-phy@47401b00 {
+		pinctrl-names = "default";
+		pinctrl-0 = <&usb1_pins>;
+		status = "disabled";	/* forever: no second USB */
+	};
+	usb@47401000 {
+		dr_mode = "host";
+	};
+	usb@47401800 {
+		status = "disabled";	/* forever: no second USB */
+		dr_mode = /*"otg"*/"host";
+	};
+};
+
+&elm {
+	status = "okay";
+};
+
+&gpmc {
+	status = "okay";
+	pinctrl-names = "default";
+	pinctrl-0 = <&nandflash_pins_s0>;
+	ranges = <0 0 0x08000000 0x20000000>;	/* CS0: NAND 512 MB (16bit) */
+	nand@0,0 {
+		compatible = "ti,omap2-nand";
+		reg = <0 0 4>; /* CS0, offset 0, IO size 4 */
+		interrupt-parent = <&gpmc>;
+		interrupts = <0 IRQ_TYPE_NONE>, /* fifoevent */
+			     <1 IRQ_TYPE_NONE>; /* termcount */
+		nand-bus-width = <16>;
+		ti,nand-ecc-opt = "bch16"; /* 16 Spansion, 8 Micron */
+		gpmc,device-nand = "true";
+		gpmc,device-width = <2>;
+		gpmc,sync-clk-ps = <0>;
+		gpmc,cs-on-ns = <0>;
+		gpmc,cs-rd-off-ns = <44>;
+		gpmc,cs-wr-off-ns = <44>;
+		gpmc,adv-on-ns = <6>;
+		gpmc,adv-rd-off-ns = <34>;
+		gpmc,adv-wr-off-ns = <44>;
+		gpmc,we-on-ns = <0>;
+		gpmc,we-off-ns = <40>;
+		gpmc,oe-on-ns = <0>;
+		gpmc,oe-off-ns = <54>;
+		gpmc,access-ns = <64>;
+		gpmc,rd-cycle-ns = <82>;
+		gpmc,wr-cycle-ns = <82>;
+		gpmc,wait-on-read = "true";
+		gpmc,wait-on-write = "true";
+		gpmc,bus-turnaround-ns = <0>;
+		gpmc,cycle2cycle-delay-ns = <0>;
+		gpmc,clk-activation-ns = <0>;
+		gpmc,wait-monitoring-ns = <0>;
+		gpmc,wr-access-ns = <40>;
+		gpmc,wr-data-mux-bus-ns = <0>;
+
+		#address-cells = <1>;
+		#size-cells = <1>;
+		elm_id = <&elm>;
+
+		/* MTD partition table */
+		partition@0 {
+			label = "NAND.SPL0";
+			reg = <0x00000000 0x000020000>;
+		};
+		partition@1 {
+			label = "NAND.SPL1";
+			reg = <0x00020000 0x00020000>;
+		};
+		partition@2 {
+			label = "NAND.SPL2";
+			reg = <0x00040000 0x00020000>;
+		};
+		partition@3 {
+			label = "NAND.SPL3";
+			reg = <0x00060000 0x00020000>;
+		};
+		partition@4 {
+			label = "NAND.splash";
+			reg = <0x00080000 0x00080000>;
+		};
+		partition@5 {
+			label = "NAND.u-boot-env";
+			reg = <0x00100000 0x00020000>;
+		};
+		partition@6 {
+			label = "NAND.u-boot-ext";
+			reg = <0x00120000 0x00020000>;
+		};
+		partition@7 {
+			label = "NAND.u-boot";
+			reg = <0x00140000 0x00140000>;
+		};
+		partition@8 {
+			label = "NAND.kernel-dts0";
+			reg = <0x00280000 0x00020000>;
+		};
+		partition@9 {
+			label = "NAND.kernel-dts1";
+			reg = <0x002A0000 0x00020000>;
+		};
+		partition@10 {
+			label = "NAND.kernel";
+			reg = <0x002C0000 0x00500000>;
+		};
+		partition@11 {
+			label = "NAND.rootfs";
+			reg = <0x007C0000 0x1F860000>;
+		};
+	};
+};
+
+&i2c_tps {
+	pinctrl-names = "default";
+	pinctrl-0 = <&i2c_tps_pins>;
+	gpios = <&gpio1 21 GPIO_ACTIVE_HIGH	/* i2c_soft0.sda, gpio1_21, 0x54 (PIN_INPUT | MUX_MODE7) */
+		 &gpio1 16 GPIO_ACTIVE_HIGH	/* i2c_soft0.scl, gpio1_16, 0x40 (PIN_OUTPUT_PULLUP | MUX_MODE7) */
+		>;
+	status = "okay";
+	tps: tps@2d {
+		reg = <0x2d>;
+	};
+	eeprom: eeprom@50 {
+		compatible = "at,24c16";
+		reg = <0x50>;
+		pagesize = <16>;
+	};
+};
+
+ /include/ "tps65910.dtsi"
+
+&tps {
+	ti,system-power-controller;
+	interrupt-parent = <&intc>;
+/*	interrupt-controller;*/
+/*	#interrupt-cells = <2>;*/
+	interrupts = <6>/*<7>*//*<7 0x2>*/;
+/*	tps65910-rtc {*/
+/*		compatible = "tps65910-rtc";*/
+/*		*/
+/*	};*/
+	ti,en-ck32k-xtal;
+	vcc1-supply = <&vbat>;
+	vcc2-supply = <&vbat>;
+	vcc3-supply = <&vbat>;
+	vcc4-supply = <&vbat>;
+	vcc5-supply = <&vbat>;
+	vcc6-supply = <&vbat>;
+	vcc7-supply = <&vbat>;
+	vccio-supply = <&vbat>;
+
+	regulators {
+		vrtc_reg: regulator@0 {
+			regulator-always-on;
+		};
+
+		vio_reg: regulator@1 {
+			regulator-always-on;
+		};
+
+		vdd1_reg: regulator@2 {
+			/* VDD_MPU voltage limits 0.95V - 1.35V with +/-4% tolerance */
+			regulator-name = "vdd_mpu";
+			regulator-min-microvolt = <912500>;
+			regulator-max-microvolt = <1350000>;
+			regulator-boot-on;
+			regulator-always-on;
+		};
+
+		vdd2_reg: regulator@3 {
+			/* VDD_CORE voltage limits 0.95V - 1.1V with +/-4% tolerance */
+			regulator-name = "vdd_core";
+			regulator-min-microvolt = <912500>;
+			regulator-max-microvolt = <1150000>;
+			regulator-boot-on;
+			regulator-always-on;
+		};
+
+		vdd3_reg: regulator@4 {
+			regulator-always-on;
+		};
+
+		vdig1_reg: regulator@5 {
+			regulator-always-on;
+		};
+
+		vdig2_reg: regulator@6 {
+			regulator-always-on;
+		};
+
+		vpll_reg: regulator@7 {
+			regulator-always-on;
+		};
+
+		vdac_reg: regulator@8 {
+			regulator-always-on;
+		};
+
+		vaux1_reg: regulator@9 {
+			regulator-always-on;
+		};
+
+		vaux2_reg: regulator@10 {
+			regulator-always-on;
+		};
+
+		vaux33_reg: regulator@11 {
+			regulator-always-on;
+		};
+
+		vmmc_reg: regulator@12 {
+			regulator-min-microvolt = <3300000>;
+			regulator-max-microvolt = <3300000>;
+			regulator-always-on;
+		};
+	};
+};
+
+&cpsw_emac0 {
+	phy_id = <&davinci_mdio>, <1>;
+	phy-mode = "mii";
+};
+
+&mac {
+	pinctrl-names = "default", "sleep";
+	pinctrl-0 = <&cpsw_default>;
+	pinctrl-1 = <&cpsw_sleep>;
+	status = "okay";
+};
+
+&davinci_mdio {
+	pinctrl-names = "default", "sleep";
+	pinctrl-0 = <&davinci_mdio_default>;
+	pinctrl-1 = <&davinci_mdio_sleep>;
+	status = "okay";
+};
+
+&sham {
+	status = "okay";
+};
+
+&aes {
+	status = "okay";
+};
+
+&mmc3 {
+	pinctrl-names = "default", "idle";
+	pinctrl-0 = <&mmc2_pins>;
+	pinctrl-1 = <&mmc2_idle>;
+	cd-gpios = <&gpio3 16 GPIO_ACTIVE_HIGH>;
+	/* or use polling: broken-cd; */
+	disable-wp;
+	cap-sd-highspeed;
+	bus-width = <4>;
+	vmmc-supply = <&vmmc_reg>;
+	/* requires &edma { ti,edma-xbar-event-map = /bits/ 16 <1 12 2 13>; }; in main DTS */
+	dmas = <&edma_xbar 12 0 1>, <&edma_xbar 13 0 2>;
+	dma-names = "tx", "rx";
+	/* try up to 50 000 000 for Tibbit. converters max:30Mz */
+	max-frequency = <20000000>;
+	status = "disabled";
+};
+
+/* required for mmc3 (4.4), no need for 4.10 */
+/* &edma {
+	ti,edma-xbar-event-map = /bits/ 16 <1 12
+					    2 13>;
+}; */
+
+&spi0 {
+    status = "disabled";
+    pinctrl-names = "default";
+    pinctrl-0 = <&spi_wifi_pinsS>;
+    ga1000v0: tpsga1000v0@0 {
+	compatible = "tpsga1000v0";
+	pinctrl-names = "default";
+	pinctrl-0 = <&spi_wifi_add_gpio>;
+	reg = <0>;
+	spi-max-frequency = <48000000>;
+	gpio-irq = <115>;	/* wifi-ext1, gpio3.19 */
+	gpio-reset = <&gpio3 20 GPIO_ACTIVE_HIGH>; /* wifi_rst gpio3.20 0x1a8 */
+	gpio-strap = <&gpio0 6 GPIO_ACTIVE_HIGH>; /* wifi_cs gpio0.06 0x154 (chipselect) */
+	status = "disabled";
+    };
+};
+
+/* for buzzer */
+&epwmss0 {  status = "okay";  };
+&ehrpwm0 {  status = "okay";  };
+/* for buzzer / */
+
