--- linux/arch/arm/boot/dts/am335x-TPPs3.dtsi	1970-01-01 03:00:00.000000000 +0300
+++ linux/arch/arm/boot/dts/am335x-TPPs3.dtsi	2014-12-16 10:12:59.961528417 +0300
@@ -0,0 +1,951 @@
+/*
+ * Tibbo Project PCB 3 for Linux (TPP3) + 4 lines
+ * Copyright (C) 2016 Tibbo Tech. - http://tibbo.com/tps/
+ * License: GPLv2 or later
+ */
+
+/ {
+	model = "AM335x TPPs3";
+	compatible = "ti,am335x-tpp3", "ti,am335x-tpp", "ti,am33xx";
+
+	aliases {
+		tty_s17 = &uart0;
+		tty_s01 = &uart1;
+		/*serial2 = &uart2;*/
+		tty_s05 = &uart3;
+		tty_s13 = &uart4;
+		tty_s09 = &uart5;
+	};
+
+	sound: sound {
+		status = "disabled";
+		compatible = "ti,da830-evm-audio";
+		ti,model = "3104 on AM335x TPP";
+		ti,audio-codec = <&tlv320aic3x>;
+		ti,mcasp-controller = <&mcasp0>;
+		ti,codec-clock-rate = /*<12000000>*/<24000000>;
+		ti,audio-routing =
+			"Headphone Jack",       "HPLOUT",
+			"Headphone Jack",       "HPROUT",
+/*			"MIC2L",                "Mic Bias",*/
+/*			"MIC2R",                "Mic Bias";*/
+			"MIC2L",                "Mic Jack",
+			"MIC2R",                "Mic Jack";
+	};
+
+};
+
+&am33xx_pinmux {
+
+	uart0_pins: pinmux_uart0_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x970, (PIN_INPUT_PULLUP | MUX_MODE0))	/* uart0_rxd.uart0_rxd */
+			AM33XX_IOPAD(0x974, (PIN_OUTPUT_PULLDOWN | MUX_MODE0))	/* uart0_txd.uart0_txd */
+		>;
+	};
+	uart0_norts_pins: uart0_norts_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x8ec, (PIN_OUTPUT | MUX_MODE7))		/* rts: gpio2.25 pin089 S17C */
+			AM33XX_IOPAD(0x970, (PIN_INPUT_PULLUP | MUX_MODE0))	/* uart0_rxd.uart0_rxd */
+			AM33XX_IOPAD(0x974, (PIN_OUTPUT_PULLDOWN | MUX_MODE0))	/* uart0_txd.uart0_txd */
+		>;
+	};
+	uart1_pins: uart1_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x978, (PIN_INPUT | MUX_MODE0))		/* cts: gpio0.12 pin012 S1D */
+			AM33XX_IOPAD(0x97c, (PIN_OUTPUT | MUX_MODE0))		/* rts: gpio0.13 pin013 S1C */
+			AM33XX_IOPAD(0x980, (PIN_INPUT | MUX_MODE0))		/* rxd: gpio0.14 pin014 S1B */
+			AM33XX_IOPAD(0x984, (PIN_OUTPUT | MUX_MODE0))		/* txd: gpio0.15 pin015 S1A */
+		>;
+	};
+	uart1_norts_pins: uart1_norts_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x978, (PIN_INPUT  | MUX_MODE0))		/* cts: gpio0.12 pin012 S1D */
+			AM33XX_IOPAD(0x97c, (PIN_OUTPUT | MUX_MODE7))		/* rts: gpio0.13 pin013 S1C */
+			AM33XX_IOPAD(0x980, (PIN_INPUT  | MUX_MODE0))		/* rxd: gpio0.14 pin014 S1B */
+			AM33XX_IOPAD(0x984, (PIN_OUTPUT | MUX_MODE0))		/* txd: gpio0.15 pin015 S1A */
+		>;
+	};
+	uart1_pins_485: uart1_pins_485 {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x9b0, (PIN_OUTPUT | MUX_MODE7))		/* fd/hd: gpio0.19 pin019 S3A */
+			AM33XX_IOPAD(0x964, (PIN_OUTPUT | MUX_MODE7))		/* mode 232/422/: gpio0.07 pin007 S3B */
+			AM33XX_IOPAD(0x9ac, (PIN_OUTPUT | MUX_MODE7))		/* dtr: gpio3.21 pin117 S3C */
+			AM33XX_IOPAD(0xa34, (PIN_INPUT  | MUX_MODE7))		/* dsr: gpio3.13 pin109 S3D */
+		>;
+	};
+	uart3_pins: uart3_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x8c8, (PIN_INPUT  | MUX_MODE6))		/* cts: gpio2.16 pin080 S5D */
+			AM33XX_IOPAD(0x8cc, (PIN_OUTPUT | MUX_MODE6))		/* rts: gpio2.17 pin081 S5C */
+			AM33XX_IOPAD(0x8f8, (PIN_INPUT  | MUX_MODE3))		/* rxd: gpio2.28 pin092 S5B */
+			AM33XX_IOPAD(0x8fc, (PIN_OUTPUT | MUX_MODE3))		/* txd: gpio2.29 pin093 S5A */
+		>;
+	};
+	uart3_norts_pins: uart3_norts_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x8c8, (PIN_INPUT  | MUX_MODE6))		/* cts: gpio2.16 pin080 S5D */
+			AM33XX_IOPAD(0x8cc, (PIN_OUTPUT | MUX_MODE7))		/* rts: gpio2.17 pin081 S5C */
+			AM33XX_IOPAD(0x8f8, (PIN_INPUT  | MUX_MODE3))		/* rxd: gpio2.28 pin092 S5B */
+			AM33XX_IOPAD(0x8fc, (PIN_OUTPUT | MUX_MODE3))		/* txd: gpio2.29 pin093 S5A */
+		>;
+	};
+	uart3_pins_485: uart3_pins_485 {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x8f4, (PIN_OUTPUT | MUX_MODE7))		/* fd/hd: gpio2.29 pin093 S7A */
+			AM33XX_IOPAD(0x9b4, (PIN_OUTPUT | MUX_MODE7))		/* mode 232/422/: gpio2.29 pin093 S7B */
+			AM33XX_IOPAD(0x944, (PIN_OUTPUT | MUX_MODE7))		/* dtr: gpio2.29 pin093 S7C */
+			AM33XX_IOPAD(0x850, (PIN_INPUT  | MUX_MODE7))		/* dsr: gpio2.29 pin093 S7D */
+		>;
+	};
+	uart4_pins: uart4_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x8d0, (PIN_INPUT  | MUX_MODE6))		/* cts: gpio0.08 pin008 S13D */
+			AM33XX_IOPAD(0x8d4, (PIN_OUTPUT | MUX_MODE6))		/* rts: gpio0.09 pin009 S13C */
+			AM33XX_IOPAD(0x968, (PIN_INPUT  | MUX_MODE1))		/* rxd: gpio1.08 pin040 S13A later change vise-versa */
+			AM33XX_IOPAD(0x96c, (PIN_OUTPUT | MUX_MODE1))		/* txd: gpio1.09 pin041 S13B later change vise-versa */
+		>;
+	};
+	uart4_norts_pins: uart4_norts_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x8d0, (PIN_INPUT  | MUX_MODE6))		/* cts: gpio0.08 pin008 S13D */
+			AM33XX_IOPAD(0x8d4, (PIN_OUTPUT | MUX_MODE7))		/* rts: gpio0.09 pin009 S13C */
+			AM33XX_IOPAD(0x968, (PIN_INPUT  | MUX_MODE1))		/* rxd: gpio1.08 pin040 S13A later change vise-versa */
+			AM33XX_IOPAD(0x96c, (PIN_OUTPUT | MUX_MODE1))		/* txd: gpio1.09 pin041 S13B later change vise-versa */
+		>;
+	};
+	uart4_pins_485: uart4_pins_485 {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x98c, (PIN_OUTPUT | MUX_MODE7))		/* fd/hd: gpio3.06 pin102 S15A */
+			AM33XX_IOPAD(0x988, (PIN_OUTPUT | MUX_MODE7))		/* mode 232/422/: gpio3.05 pin101 S15B */
+			AM33XX_IOPAD(0x990, (PIN_OUTPUT | MUX_MODE7))		/* dtr: gpio3.14 pin110 S15C */
+			AM33XX_IOPAD(0x99c, (PIN_INPUT  | MUX_MODE7))		/* dsr: gpio3.17 pin113 S15D */
+		>;
+	};
+	uart5_pins: uart5_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x8d8, (PIN_INPUT  | MUX_MODE6))		/* cts: gpio0.10 pin010 S9D */
+			AM33XX_IOPAD(0x8dc, (PIN_OUTPUT | MUX_MODE6))		/* rts: gpio0.11 pin011 S9C */
+			AM33XX_IOPAD(0x8c4, (PIN_INPUT  | MUX_MODE4))		/* rxd: gpio2.15 pin079 S9B */
+			AM33XX_IOPAD(0x8c0, (PIN_OUTPUT | MUX_MODE4))		/* txd: gpio2.14 pin078 S9A */
+		>;
+	};
+	uart5_norts_pins: uart5_norts_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x8d8, (PIN_INPUT  | MUX_MODE6))		/* cts: gpio0.10 pin010 S9D */
+			AM33XX_IOPAD(0x8dc, (PIN_OUTPUT | MUX_MODE7))		/* rts: gpio0.11 pin011 S9C */
+			AM33XX_IOPAD(0x8c4, (PIN_INPUT  | MUX_MODE4))		/* rxd: gpio2.15 pin079 S9B */
+			AM33XX_IOPAD(0x8c0, (PIN_OUTPUT | MUX_MODE4))		/* txd: gpio2.14 pin078 S9A */
+		>;
+	};
+	uart5_pins_485: uart5_pins_485 {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x8f0, (PIN_OUTPUT | MUX_MODE7))		/* fd/hd: gpio2.26 pin078 S11A */
+			AM33XX_IOPAD(0x85c, (PIN_OUTPUT | MUX_MODE7))		/* mode 232/422/485: gpio1.23 pin078 S11B */
+			AM33XX_IOPAD(0x868, (PIN_OUTPUT | MUX_MODE7))		/* dtr: gpio1.26 pin078 S11C */
+			AM33XX_IOPAD(0x9a0, (PIN_INPUT  | MUX_MODE7))		/* dsr: gpio3.18 pin078 S11D */
+		>;
+	};
+
+	i2c_s01_pins: i2c_s01_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x984, (PIN_INPUT | MUX_MODE7))	/* has to be 3f|2f S01A */
+			AM33XX_IOPAD(0x980, (PIN_INPUT | MUX_MODE7))	/* has to be 3f|2f S01B */
+		>;
+	};
+	i2c_s03_pins: i2c_s03_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x9b0, (PIN_INPUT | MUX_MODE7))	/* has to be 3f|2f S03A */
+			AM33XX_IOPAD(0x964, (PIN_INPUT | MUX_MODE7))	/* has to be 3f|2f S03B */
+		>;
+	};
+	i2c_s05_pins: i2c_s05_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x8fc, (PIN_INPUT | MUX_MODE7))	/* has to be 3f|2f S05A */
+			AM33XX_IOPAD(0x8f8, (PIN_INPUT | MUX_MODE7))	/* has to be 3f|2f S05B */
+		>;
+	};
+	i2c_s07_pins: i2c_s07_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x8f4, (PIN_INPUT | MUX_MODE7))	/* has to be 3f|2f S07A */
+			AM33XX_IOPAD(0x9b4, (PIN_INPUT | MUX_MODE7))	/* has to be 3f|2f S07B */
+		>;
+	};
+	i2c_s09_pins: i2c_s09_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x8c0, (PIN_INPUT | MUX_MODE7))	/* has to be 3f|2f S09A */
+			AM33XX_IOPAD(0x8c4, (PIN_INPUT | MUX_MODE7))	/* has to be 3f|2f S09B */
+		>;
+	};
+	i2c_s11_pins: i2c_s11_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x8f0, (PIN_INPUT | MUX_MODE7))	/* has to be 3f|2f S11A */
+			AM33XX_IOPAD(0x85c, (PIN_INPUT | MUX_MODE7))	/* has to be 3f|2f S11B */
+		>;
+	};
+	i2c_s13_pins: i2c_s13_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x96c, (PIN_INPUT | MUX_MODE7))	/* has to be 3f|2f S13A */
+			AM33XX_IOPAD(0x968, (PIN_INPUT | MUX_MODE7))	/* has to be 3f|2f S13B */
+		>;
+	};
+	i2c_s15_pins: i2c_s15_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x98c, (PIN_INPUT | MUX_MODE7))	/* has to be 3f|2f S15A */
+			AM33XX_IOPAD(0x988, (PIN_INPUT | MUX_MODE7))	/* has to be 3f|2f S15B */
+		>;
+	};
+	i2c_s17_pins: i2c_s17_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x974, (PIN_INPUT | MUX_MODE7))	/* has to be 3f|2f S17A */
+			AM33XX_IOPAD(0x970, (PIN_INPUT | MUX_MODE7))	/* has to be 3f|2f S17B */
+		>;
+	};
+	i2c_s19_pins: i2c_s19_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x904, (PIN_INPUT | MUX_MODE7))	/* has to be 3f|2f S19A */
+			AM33XX_IOPAD(0x900, (PIN_INPUT | MUX_MODE7))	/* has to be 3f|2f S19B */
+		>;
+	};
+	i2c_s21_pins: i2c_s21_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x88c, (PIN_INPUT | MUX_MODE7))	/* has to be 3f|2f S21A */
+			AM33XX_IOPAD(0x888, (PIN_INPUT | MUX_MODE7))	/* has to be 3f|2f S21B */
+		>;
+	};
+	i2c_s23_pins: i2c_s23_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x844, (PIN_INPUT | MUX_MODE7))	/* has to be 3f|2f S23A */
+			AM33XX_IOPAD(0x848, (PIN_INPUT | MUX_MODE7))	/* has to be 3f|2f S23B */
+		>;
+	};
+	i2c_s25_pins: i2c_s25_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x8e8, (PIN_INPUT | MUX_MODE7))	/* has to be 3f|2f S25A */
+			AM33XX_IOPAD(0x8e0, (PIN_INPUT | MUX_MODE7))	/* has to be 3f|2f S25B */
+		>;
+	};
+
+	spi_s01_pins: spi_s01_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x984, (PIN_OUTPUT_PULLUP | MUX_MODE7))	/* gpio0.15 pin097 S01A cs */
+			AM33XX_IOPAD(0x980, (PIN_OUTPUT_PULLUP | MUX_MODE7))	/* gpio0.14 pin096 S01B clk */
+			AM33XX_IOPAD(0x97c, (PIN_OUTPUT_PULLUP | MUX_MODE7))	/* gpio0.13 pin095 S01C mosi */
+			AM33XX_IOPAD(0x978, (PIN_INPUT_PULLUP  | MUX_MODE7))	/* gpio0.12 pin094 S01D miso */
+		>;
+	};
+	spi_s03_pins: spi_s03_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x9b0, (PIN_OUTPUT_PULLUP | MUX_MODE7))	/* gpio0.19 pin0xx S03A cs */
+			AM33XX_IOPAD(0x964, (PIN_OUTPUT_PULLUP | MUX_MODE7))	/* gpio0.07 pin0xx S03B clk */
+			AM33XX_IOPAD(0x9ac, (PIN_OUTPUT_PULLUP | MUX_MODE7))	/* gpio3.21 pin0xx S03C mosi */
+			AM33XX_IOPAD(0xa34, (PIN_INPUT_PULLUP  | MUX_MODE7))	/* gpio3.13 pin141 S03D miso */
+		>;
+	};
+	spi_s05_pins: spi_s05_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x8fc, (PIN_OUTPUT_PULLUP | MUX_MODE7))	/* gpio2.29 pin093 S05A cs */
+			AM33XX_IOPAD(0x8f8, (PIN_OUTPUT_PULLUP | MUX_MODE7))	/* gpio2.28 pin092 S05B clk */
+			AM33XX_IOPAD(0x8cc, (PIN_OUTPUT_PULLUP | MUX_MODE7))	/* gpio2.17 pin081 S05C mosi */
+			AM33XX_IOPAD(0x8c8, (PIN_INPUT_PULLUP  | MUX_MODE7))	/* gpio2.16 pin080 S05D miso */
+		>;
+	};
+	spi_s07_pins: spi_s07_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x8f4, (PIN_OUTPUT_PULLUP | MUX_MODE7))	/* gpio2.27 pin061 S07A cs */
+			AM33XX_IOPAD(0x9b4, (PIN_OUTPUT_PULLUP | MUX_MODE7))	/* gpio0.20 pin0xx S07B clk */
+			AM33XX_IOPAD(0x944, (PIN_OUTPUT_PULLUP | MUX_MODE7))	/* gpio0.29 pin0xx S07C mosi */
+			AM33XX_IOPAD(0x850, (PIN_INPUT_PULLUP  | MUX_MODE7))	/* gpio1.20 pin0xx S07D miso */
+		>;
+	};
+	spi_s09_pins: spi_s09_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x8c0, (PIN_OUTPUT_PULLUP | MUX_MODE7))	/* gpio2.14 pin048 S09A cs */
+			AM33XX_IOPAD(0x8c4, (PIN_OUTPUT_PULLUP | MUX_MODE7))	/* gpio2.15 pin049 S09B clk */
+			AM33XX_IOPAD(0x8dc, (PIN_OUTPUT_PULLUP | MUX_MODE7))	/* gpio0.11 pin055 S09C mosi */
+			AM33XX_IOPAD(0x8d8, (PIN_INPUT_PULLUP  | MUX_MODE7))	/* gpio0.10 pin054 S09D miso */
+		>;
+	};
+	spi_s11_pins: spi_s11_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x8f0, (PIN_OUTPUT_PULLUP | MUX_MODE7))	/* gpio2.26 pin0xx S11A cs */
+			AM33XX_IOPAD(0x85c, (PIN_OUTPUT_PULLUP | MUX_MODE7))	/* gpio1.23 pin0xx S11B clk */
+			AM33XX_IOPAD(0x868, (PIN_OUTPUT_PULLUP | MUX_MODE7))	/* gpio1.26 pin0xx S11C mosi */
+			AM33XX_IOPAD(0x9a0, (PIN_INPUT_PULLUP  | MUX_MODE7))	/* gpio3.18 pin0xx S11D miso */
+		>;
+	};
+	spi_s13_pins: spi_s13_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x96c, (PIN_OUTPUT_PULLUP | MUX_MODE7))	/* gpio1.09 pin091 S13A cs */
+			AM33XX_IOPAD(0x968, (PIN_OUTPUT_PULLUP | MUX_MODE7))	/* gpio1.08 pin090 S13B clk */
+			AM33XX_IOPAD(0x8d5, (PIN_OUTPUT_PULLUP | MUX_MODE7))	/* gpio0.09 pin053 S13C mosi */
+			AM33XX_IOPAD(0x8d0, (PIN_INPUT_PULLUP  | MUX_MODE7))	/* gpio0.08 pin052 S13D miso */
+		>;
+	};
+	spi_s15_pins: spi_s15_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x98c, (PIN_OUTPUT_PULLUP | MUX_MODE7))	/* gpio3.06 pin0xx S15A cs */
+			AM33XX_IOPAD(0x988, (PIN_OUTPUT_PULLUP | MUX_MODE7))	/* gpio3.05 pin0xx S15B clk */
+			AM33XX_IOPAD(0x990, (PIN_OUTPUT_PULLUP | MUX_MODE7))	/* gpio3.14 pin0xx S15C mosi */
+			AM33XX_IOPAD(0x99c, (PIN_INPUT_PULLUP  | MUX_MODE7))	/* gpio3.17 pin0xx S15D miso */
+		>;
+	};
+	spi_s17_pins: spi_s17_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x974, (PIN_OUTPUT_PULLUP | MUX_MODE7))	/* gpio1.11 pin0xx S17A cs */
+			AM33XX_IOPAD(0x970, (PIN_OUTPUT_PULLUP | MUX_MODE7))	/* gpio1.10 pin0xx S17B clk */
+			AM33XX_IOPAD(0x8ec, (PIN_OUTPUT_PULLUP | MUX_MODE7))	/* gpio2.25 pin0xx S17C mosi */
+			AM33XX_IOPAD(0x8e4, (PIN_INPUT_PULLUP  | MUX_MODE7))	/* gpio2.23 pin0xx S17D miso */
+		>;
+	};
+	spi_s19_pins: spi_s19_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x904, (PIN_OUTPUT_PULLUP | MUX_MODE7))	/* gpio2.31 pin0xx S19A cs */
+			AM33XX_IOPAD(0x900, (PIN_OUTPUT_PULLUP | MUX_MODE7))	/* gpio2.30 pin0xx S19B clk */
+			AM33XX_IOPAD(0x884, (PIN_OUTPUT_PULLUP | MUX_MODE7))	/* gpio1.31 pin0xx S19C mosi */
+			AM33XX_IOPAD(0x880, (PIN_INPUT_PULLUP  | MUX_MODE7))	/* gpio1.30 pin0xx S19D miso */
+		>;
+	};
+	spi_s21_pins: spi_s21_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x88c, (PIN_OUTPUT_PULLUP | MUX_MODE7))	/* gpio2.01 pin0xx S21A cs */
+			AM33XX_IOPAD(0x888, (PIN_OUTPUT_PULLUP | MUX_MODE7))	/* gpio2.00 pin0xx S21B clk */
+			AM33XX_IOPAD(0x998, (PIN_OUTPUT_PULLUP | MUX_MODE7))	/* gpio3.16 pin0xx S21C mosi */
+			AM33XX_IOPAD(0x95c, (PIN_INPUT_PULLUP  | MUX_MODE7))	/* gpio0.05 pin0xx S21D miso */
+		>;
+	};
+	spi_s23_pins: spi_s23_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x844, (PIN_OUTPUT_PULLUP | MUX_MODE7))	/* gpio1.17 pin0xx S23A cs */
+			AM33XX_IOPAD(0x848, (PIN_OUTPUT_PULLUP | MUX_MODE7))	/* gpio1.18 pin0xx S23B clk */
+			AM33XX_IOPAD(0x84c, (PIN_OUTPUT_PULLUP | MUX_MODE7))	/* gpio1.19 pin0xx S23C mosi */
+			AM33XX_IOPAD(0x878, (PIN_INPUT_PULLUP  | MUX_MODE7))	/* gpio1.28 pin0xx S23D miso */
+		>;
+	};
+
+	dcan0_pins: dcan0_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x970, (PIN_OUTPUT | MUX_MODE2))	/* gpio1.10 pin42  S17B tx */
+			AM33XX_IOPAD(0x974, (PIN_INPUT  | MUX_MODE2))	/* gpio1.11 pin43  S17A rx */
+		>;
+	};
+	dcan1_pins: dcan1_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x900, (PIN_OUTPUT | MUX_MODE4))	/* gpio2.30 pin94  S19B tx */
+			AM33XX_IOPAD(0x904, (PIN_INPUT  | MUX_MODE4))	/* gpio2.31 pin95  S19A rx */
+		>;
+	};
+
+	sound_pins: sound_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x8dc, (PIN_INPUT  | MUX_MODE3)) 	/* mcasp0_ahclkx S9C */
+			AM33XX_IOPAD(0x8c4, (PIN_INPUT  | MUX_MODE3))	/* mcasp0_fsx	 S9B */
+			AM33XX_IOPAD(0x8c0, (PIN_INPUT  | MUX_MODE3))	/* mcasp_aclkx ? S9A */
+			AM33XX_IOPAD(0x9a0, (PIN_OUTPUT  | MUX_MODE2))	/* mcasp_axr2    S7D */
+			AM33XX_IOPAD(0x868, (PIN_INPUT | MUX_MODE6))	/* mcasp_axr0    S7B */
+		>;
+	};
+
+};
+
+/ {
+
+	i2c_s01: i2c-s01 { /* S1A, S1B, custom */
+		compatible = "i2c-gpio";
+		i2c-gpio,delay-us = <0>;	/* ~100 kHz */
+		#address-cells = <1>;
+		#size-cells = <0>;
+		status = "disabled";
+		pinctrl-names = "default";
+		pinctrl-0 = <&i2c_s01_pins>;
+		gpios = <&gpio0 14 GPIO_ACTIVE_HIGH	/* i2c_soft2.sda, gpio0_15, S1B, 0x180 (PIN_OUTPUT_PULLUP | MUX_MODE7) */
+			 &gpio0 15 GPIO_ACTIVE_HIGH	/* i2c_soft2.scl, gpio0_14, S1A, 0x184 (PIN_INPUT | MUX_MODE7) */
+			>;
+	};
+	i2c_s03: i2c-s03 { /* S3A, S3B, custom */
+		compatible = "i2c-gpio";
+		i2c-gpio,delay-us = <0>;	/* ~100 kHz */
+		#address-cells = <1>;
+		#size-cells = <0>;
+		status = "disabled";
+		pinctrl-names = "default";
+		pinctrl-0 = <&i2c_s03_pins>;
+		gpios = <&gpio0 7  GPIO_ACTIVE_HIGH	/* i2c_soft3.sda, gpio0_07, S3B, 0x164 (PIN_OUTPUT_PULLUP | MUX_MODE7) */
+			 &gpio0 19 GPIO_ACTIVE_HIGH	/* i2c_soft3.scl, gpio0_19, S3A, 0x1b0 (PIN_INPUT | MUX_MODE7) */
+			>;
+	};
+	i2c_s05: i2c-s05 { /* S5A, S5B, custom */
+		compatible = "i2c-gpio";
+		i2c-gpio,delay-us = <0>;	/* ~100 kHz */
+		#address-cells = <1>;
+		#size-cells = <0>;
+		status = "disabled";
+		pinctrl-names = "default";
+		pinctrl-0 = <&i2c_s05_pins>;
+		gpios = <&gpio2 28 GPIO_ACTIVE_HIGH	/* i2c_soft2.sda, gpio2_28, S5B, 0xf8 (PIN_OUTPUT_PULLUP | MUX_MODE7) */
+			 &gpio2 29 GPIO_ACTIVE_HIGH	/* i2c_soft2.scl, gpio2_29, S5A, 0xfc (PIN_INPUT | MUX_MODE7) */
+			>;
+	};
+	i2c_s07: i2c-s07 { /* S7A, S7B, custom */
+		compatible = "i2c-gpio";
+		i2c-gpio,delay-us = <0>;	/* ~100 kHz */
+		#address-cells = <1>;
+		#size-cells = <0>;
+		status = "disabled";
+		pinctrl-names = "default";
+		pinctrl-0 = <&i2c_s07_pins>;
+		gpios = <&gpio0 20 GPIO_ACTIVE_HIGH	/* i2c_soft3.sda, gpio0_20, S7B, 0x1b4 (PIN_OUTPUT_PULLUP | MUX_MODE7) */
+			 &gpio2 27 GPIO_ACTIVE_HIGH	/* i2c_soft3.scl, gpio2_27, S7A, 0xf4 (PIN_INPUT | MUX_MODE7) */
+			>;
+	};
+	i2c_s09: i2c-s09 { /* S9A, S9B, custom */
+		compatible = "i2c-gpio";
+		i2c-gpio,delay-us = <0>;	/* ~100 kHz */
+		#address-cells = <1>;
+		#size-cells = <0>;
+		status = "disabled";
+		pinctrl-names = "default";
+		pinctrl-0 = <&i2c_s09_pins>;
+		gpios = <&gpio2 15 GPIO_ACTIVE_HIGH	/* i2c_soft.sda, gpio2_15, S9B, 0xc4 (PIN_OUTPUT_PULLUP | MUX_MODE7) */
+			 &gpio2 14 GPIO_ACTIVE_HIGH	/* i2c_soft.scl, gpio2_14, S9A, 0xc0 (PIN_INPUT | MUX_MODE7) */
+			>;
+	};
+	i2c_s11: i2c-s11 { /* S11A, S11B, custom */
+		compatible = "i2c-gpio";
+		i2c-gpio,delay-us = <0>;	/* ~100 kHz */
+		#address-cells = <1>;
+		#size-cells = <0>;
+		status = "disabled";
+		pinctrl-names = "default";
+		pinctrl-0 = <&i2c_s11_pins>;
+		gpios = <&gpio1 23 GPIO_ACTIVE_HIGH	/* i2c_soft.sda, gpio1_23, S11B, 0x5c (PIN_OUTPUT_PULLUP | MUX_MODE7) */
+			 &gpio2 26 GPIO_ACTIVE_HIGH	/* i2c_soft.scl, gpio2_27, S11A, 0xf0 (PIN_INPUT | MUX_MODE7) */
+			>;
+		tlv320aic3x: tlv320aic3x@18 {
+			compatible = "ti,tlv320aic3104";
+			reg = <0x18>;
+			status = "disabled";
+			ai3x-micbias-vg = <2>;
+			gpio-reset = <&gpio0 10 GPIO_ACTIVE_HIGH>;
+
+			/* Regulators */
+			AVDD-supply = <&vaux2_reg>;
+			IOVDD-supply = <&vaux2_reg>;
+			DRVDD-supply = <&vio_reg>;
+			DVDD-supply = <&vio_reg>;
+		};
+	};
+	i2c_s13: i2c-s13 { /* S13A, S13B, custom */
+		compatible = "i2c-gpio";
+		i2c-gpio,delay-us = <0>;	/* ~100 kHz */
+		#address-cells = <1>;
+		#size-cells = <0>;
+		status = "disabled";
+		pinctrl-names = "default";
+		pinctrl-0 = <&i2c_s13_pins>;
+		gpios = <&gpio1 8 GPIO_ACTIVE_HIGH	/* i2c_soft.sda, gpio1_8, S13B, 0x168 (PIN_OUTPUT_PULLUP | MUX_MODE7) */
+			 &gpio1 9 GPIO_ACTIVE_HIGH	/* i2c_soft.scl, gpio1_9, S13A, 0x16c (PIN_INPUT | MUX_MODE7) */
+			>;
+	};
+	i2c_s15: i2c-s15 { /* S15A, S15B, custom */
+		compatible = "i2c-gpio";
+		i2c-gpio,delay-us = <0>;	/* ~100 kHz */
+		#address-cells = <1>;
+		#size-cells = <0>;
+		status = "disabled";
+		pinctrl-names = "default";
+		pinctrl-0 = <&i2c_s15_pins>;
+		gpios = <&gpio3 5 GPIO_ACTIVE_HIGH	/* i2c_soft.sda, gpio3_5, S15B, 0x188 (PIN_OUTPUT_PULLUP | MUX_MODE7) */
+			 &gpio3 6 GPIO_ACTIVE_HIGH	/* i2c_soft.scl, gpio3_6, S15A, 0x18c (PIN_INPUT | MUX_MODE7) */
+			>;
+	};
+	i2c_s17: i2c-s17 { /* S17A, S17B, custom */
+		compatible = "i2c-gpio";
+		i2c-gpio,delay-us = <0>;	/* ~100 kHz */
+		#address-cells = <1>;
+		#size-cells = <0>;
+		status = "disabled";
+		pinctrl-names = "default";
+		pinctrl-0 = <&i2c_s17_pins>;
+		gpios = <&gpio1 10 GPIO_ACTIVE_HIGH	/* i2c_soft.sda, gpio1_10, S17B, 0x170 (PIN_OUTPUT_PULLUP | MUX_MODE7) */
+			 &gpio1 11 GPIO_ACTIVE_HIGH	/* i2c_soft.scl, gpio1_11, S17A, 0x174 (PIN_INPUT | MUX_MODE7) */
+			>;
+	};
+	i2c_s19: i2c-s19 { /* S19A, S19B, custom */
+		compatible = "i2c-gpio";
+		i2c-gpio,delay-us = <0>;	/* ~100 kHz */
+		#address-cells = <1>;
+		#size-cells = <0>;
+		status = "disabled";
+		pinctrl-names = "default";
+		pinctrl-0 = <&i2c_s19_pins>;
+		gpios = <&gpio2 30 GPIO_ACTIVE_HIGH	/* i2c_soft.sda, gpio2_30, S19B, 0x100 (PIN_OUTPUT_PULLUP | MUX_MODE7) */
+			 &gpio2 31 GPIO_ACTIVE_HIGH	/* i2c_soft.scl, gpio2_31, S19A, 0x104 (PIN_INPUT | MUX_MODE7) */
+			>;
+	};
+	i2c_s21: i2c-s21 { /* S21A, S21B, custom */
+		compatible = "i2c-gpio";
+		i2c-gpio,delay-us = <0>;	/* ~100 kHz */
+		#address-cells = <1>;
+		#size-cells = <0>;
+		status = "disabled";
+		pinctrl-names = "default";
+		pinctrl-0 = <&i2c_s21_pins>;
+		gpios = <&gpio2 0 GPIO_ACTIVE_HIGH	/* i2c_soft.sda, gpio2_0, S21B, 0x88 (PIN_OUTPUT_PULLUP | MUX_MODE7) */
+			 &gpio2 1 GPIO_ACTIVE_HIGH	/* i2c_soft.scl, gpio2_1, S21A, 0x8c (PIN_INPUT | MUX_MODE7) */
+			>;
+	};
+	i2c_s23: i2c-s23 { /* S23A, S23B, custom */
+		compatible = "i2c-gpio";
+		i2c-gpio,delay-us = <0>;	/* ~100 kHz */
+		#address-cells = <1>;
+		#size-cells = <0>;
+		status = "disabled";
+		pinctrl-names = "default";
+		pinctrl-0 = <&i2c_s23_pins>;
+		gpios = <&gpio1 18 GPIO_ACTIVE_HIGH	/* i2c_soft.sda, gpio1_18, S23B, 0x48 (PIN_OUTPUT_PULLUP | MUX_MODE7) */
+			 &gpio1 17 GPIO_ACTIVE_HIGH	/* i2c_soft.scl, gpio1_17, S23A, 0x44 (PIN_INPUT | MUX_MODE7) */
+			>;
+	};
+	i2c_s25: i2c-s25 { /* S25A, S25B, custom */
+		compatible = "i2c-gpio";
+		i2c-gpio,delay-us = <0>;	/* ~100 kHz */
+		#address-cells = <1>;
+		#size-cells = <0>;
+		status = "disabled";
+		pinctrl-names = "default";
+		pinctrl-0 = <&i2c_s25_pins>;
+		gpios = <&gpio2 24 GPIO_ACTIVE_HIGH	/* i2c_soft.sda, gpio2_24, S25B, 0xe0 (PIN_OUTPUT_PULLUP | MUX_MODE7) */
+			 &gpio2 22 GPIO_ACTIVE_HIGH	/* i2c_soft.scl, gpio2_22, S25A, 0x21c (PIN_INPUT | MUX_MODE7) */
+			>;
+	};
+
+	spi_s01: spi-s01 {
+		compatible = "spi-gpio";
+		#address-cells = <0x1>;
+		#size-cells = <0x0>;
+		cs-gpios  = <&gpio0 15 GPIO_ACTIVE_HIGH>; /* S01A gpio0.15 0x184 */
+		gpio-sck  = <&gpio0 14 GPIO_ACTIVE_HIGH>; /* S01B gpio0.14 0x180 */
+		gpio-mosi = <&gpio0 13 GPIO_ACTIVE_HIGH>; /* S01C gpio0.13 0x17c */
+		gpio-miso = <&gpio0 12 GPIO_ACTIVE_HIGH>; /* S01D gpio0.12 0x178 */
+		num-chipselects = <1>;
+		pinctrl-names = "default";
+		pinctrl-0 = <&spi_s01_pins>;
+		status = "disabled";
+		spi@0 {
+			compatible = "tps,spigen";
+			spi-max-frequency = <57600000>;
+			reg = <0>;
+		};
+	};
+	spi_s03: spi-s03 {
+		compatible = "spi-gpio";
+		#address-cells = <0x1>;
+		#size-cells = <0x0>;
+		cs-gpios  = <&gpio0 19 GPIO_ACTIVE_HIGH>; /* S03A gpio0.19 0x1b0 */
+		gpio-sck  = <&gpio0 7  GPIO_ACTIVE_HIGH>; /* S03B gpio0.07 0x164 */
+		gpio-mosi = <&gpio3 21 GPIO_ACTIVE_HIGH>; /* S03C gpio3.21 0x1ac */
+		gpio-miso = <&gpio3 13 GPIO_ACTIVE_HIGH>; /* S03D gpio3.13 0x234 */
+		num-chipselects = <1>;
+		pinctrl-names = "default";
+		pinctrl-0 = <&spi_s03_pins>;
+		status = "disabled";
+		spi@0 {
+			compatible = "tps,spigen";
+			spi-max-frequency = <57600000>;
+			reg = <0>;
+		};
+	};
+	spi_s05: spi-s05 {
+		compatible = "spi-gpio";
+		#address-cells = <0x1>;
+		#size-cells = <0x0>;
+		cs-gpios  = <&gpio2 29 GPIO_ACTIVE_HIGH>; /* S05A gpio2.29 0xfc */
+		gpio-sck  = <&gpio2 28 GPIO_ACTIVE_HIGH>; /* S05B gpio2.28 0xf8 */
+		gpio-mosi = <&gpio2 17 GPIO_ACTIVE_HIGH>; /* S05C gpio2.17 0xcc */
+		gpio-miso = <&gpio2 16 GPIO_ACTIVE_HIGH>; /* S05D gpio2.16 0xc8 */
+		num-chipselects = <1>;
+		pinctrl-names = "default";
+		pinctrl-0 = <&spi_s05_pins>;
+		status = "disabled";
+		spi@0 {
+			compatible = "tps,spigen";
+			spi-max-frequency = <57600000>;
+			reg = <0>;
+		};
+	};
+	spi_s07: spi-s07 {
+		compatible = "spi-gpio";
+		#address-cells = <0x1>;
+		#size-cells = <0x0>;
+		cs-gpios  = <&gpio2 27 GPIO_ACTIVE_HIGH>; /* S07A gpio2.27 0xf4 */
+		gpio-sck  = <&gpio0 20 GPIO_ACTIVE_HIGH>; /* S07B gpio0.20 0x1b4 */
+		gpio-mosi = <&gpio0 29 GPIO_ACTIVE_HIGH>; /* S07C gpio0.29 0x144 */
+		gpio-miso = <&gpio1 20 GPIO_ACTIVE_HIGH>; /* S07D gpio1.20 0x50 */
+		num-chipselects = <1>;
+		pinctrl-names = "default";
+		pinctrl-0 = <&spi_s07_pins>;
+		status = "disabled";
+		spi@0 {
+			compatible = "tps,spigen";
+			spi-max-frequency = <57600000>;
+			reg = <0>;
+		};
+	};
+	spi_s09: spi-s09 {
+		compatible = "spi-gpio";
+		#address-cells = <0x1>;
+		#size-cells = <0x0>;
+		cs-gpios  = <&gpio2 14 GPIO_ACTIVE_HIGH>; /* S09A gpio2.14 0xc0 */
+		gpio-sck  = <&gpio2 15 GPIO_ACTIVE_HIGH>; /* S09B gpio2.15 0xc4 */
+		gpio-mosi = <&gpio0 11 GPIO_ACTIVE_HIGH>; /* S09C gpio0.11 0xdc */
+		gpio-miso = <&gpio0 10 GPIO_ACTIVE_HIGH>; /* S09D gpio0.10 0xd8 */
+		num-chipselects = <1>;
+		pinctrl-names = "default";
+		pinctrl-0 = <&spi_s09_pins>;
+		status = "disabled";
+		spi@0 {
+			compatible = "tps,spigen";
+			spi-max-frequency = <57600000>;
+			reg = <0>;
+		};
+	};
+	spi_s11: spi-s11 {
+		compatible = "spi-gpio";
+		#address-cells = <0x1>;
+		#size-cells = <0x0>;
+		cs-gpios  = <&gpio2 26 GPIO_ACTIVE_HIGH>; /* S11A gpio2.26 0xf0 */
+		gpio-sck  = <&gpio1 23 GPIO_ACTIVE_HIGH>; /* S11B gpio1.23 0x5c */
+		gpio-mosi = <&gpio1 26 GPIO_ACTIVE_HIGH>; /* S11C gpio1.26 0x68 */
+		gpio-miso = <&gpio3 18 GPIO_ACTIVE_HIGH>; /* S11D gpio3.18 0x1a0 */
+		num-chipselects = <1>;
+		pinctrl-names = "default";
+		pinctrl-0 = <&spi_s11_pins>;
+		status = "disabled";
+		spi@0 {
+			compatible = "tps,spigen";
+			spi-max-frequency = <57600000>;
+			reg = <0>;
+		};
+	};
+	spi_s13: spi-s13 {
+		compatible = "spi-gpio";
+		#address-cells = <0x1>;
+		#size-cells = <0x0>;
+		cs-gpios  = <&gpio1 9 GPIO_ACTIVE_HIGH>; /* S13A gpio1.09 0x16c */
+		gpio-sck  = <&gpio1 8 GPIO_ACTIVE_HIGH>; /* S13B gpio1.08 0x168 */
+		gpio-mosi = <&gpio0 9 GPIO_ACTIVE_HIGH>; /* S13C gpio0.09 0xd4 */
+		gpio-miso = <&gpio0 8 GPIO_ACTIVE_HIGH>; /* S13D gpio0.08 0xd0 */
+		num-chipselects = <1>;
+		pinctrl-names = "default";
+		pinctrl-0 = <&spi_s13_pins>;
+		status = "disabled";
+		spi@0 {
+			compatible = "tps,spigen";
+			spi-max-frequency = <57600000>;
+			reg = <0>;
+		};
+	};
+	spi_s15: spi-s15 {
+		compatible = "spi-gpio";
+		#address-cells = <0x1>;
+		#size-cells = <0x0>;
+		cs-gpios  = <&gpio3  6 GPIO_ACTIVE_HIGH>; /* S15A gpio3.06 0x18c */
+		gpio-sck  = <&gpio3  5 GPIO_ACTIVE_HIGH>; /* S15B gpio3.05 0x188 */
+		gpio-mosi = <&gpio3 14 GPIO_ACTIVE_HIGH>; /* S15C gpio3.14 0x190 */
+		gpio-miso = <&gpio3 17 GPIO_ACTIVE_HIGH>; /* S15D gpio3.17 0x19c */
+		num-chipselects = <1>;
+		pinctrl-names = "default";
+		pinctrl-0 = <&spi_s15_pins>;
+		status = "disabled";
+		spi@0 {
+			compatible = "tps,spigen";
+			spi-max-frequency = <57600000>;
+			reg = <0>;
+		};
+	};
+	spi_s17: spi-s17 {
+		compatible = "spi-gpio";
+		#address-cells = <0x1>;
+		#size-cells = <0x0>;
+		cs-gpios  = <&gpio1 11 GPIO_ACTIVE_HIGH>; /* S17A gpio1.11 0x174 */
+		gpio-sck  = <&gpio1 10 GPIO_ACTIVE_HIGH>; /* S17B gpio1.10 0x170 */
+		gpio-mosi = <&gpio2 25 GPIO_ACTIVE_HIGH>; /* S17C gpio2.25 0xec */
+		gpio-miso = <&gpio2 23 GPIO_ACTIVE_HIGH>; /* S17D gpio2.23 0xe4 */
+		num-chipselects = <1>;
+		pinctrl-names = "default";
+		pinctrl-0 = <&spi_s17_pins>;
+		status = "disabled";
+		spi@0 {
+			compatible = "tps,spigen";
+			spi-max-frequency = <57600000>;
+			reg = <0>;
+		};
+	};
+	spi_s19: spi-s19 {
+		compatible = "spi-gpio";
+		#address-cells = <0x1>;
+		#size-cells = <0x0>;
+		cs-gpios  = <&gpio2 31 GPIO_ACTIVE_HIGH>; /* S19A gpio2.31 0x104 */
+		gpio-sck  = <&gpio2 30 GPIO_ACTIVE_HIGH>; /* S19B gpio2.30 0x100 */
+		gpio-mosi = <&gpio1 31 GPIO_ACTIVE_HIGH>; /* S19C gpio1.31 0x84 */
+		gpio-miso = <&gpio1 30 GPIO_ACTIVE_HIGH>; /* S19D gpio1.30 0x80 */
+		num-chipselects = <1>;
+		pinctrl-names = "default";
+		pinctrl-0 = <&spi_s19_pins>;
+		status = "disabled";
+		spi@0 {
+			compatible = "tps,spigen";
+			spi-max-frequency = <57600000>;
+			reg = <0>;
+		};
+	};
+	spi_s21: spi-s21 {
+		compatible = "spi-gpio";
+		#address-cells = <0x1>;
+		#size-cells = <0x0>;
+		cs-gpios  = <&gpio2  1 GPIO_ACTIVE_HIGH>; /* S21A gpio2.01 0x8c */
+		gpio-sck  = <&gpio2  0 GPIO_ACTIVE_HIGH>; /* S21B gpio2.00 0x88 */
+		gpio-mosi = <&gpio3 16 GPIO_ACTIVE_HIGH>; /* S21C gpio3.16 0x198 */
+		gpio-miso = <&gpio0  5 GPIO_ACTIVE_HIGH>; /* S21D gpio0.05 0x15c */
+		num-chipselects = <1>;
+		pinctrl-names = "default";
+		pinctrl-0 = <&spi_s21_pins>;
+		status = "disabled";
+		spi@0 {
+			compatible = "tps,spigen";
+			spi-max-frequency = <57600000>;
+			reg = <0>;
+		};
+	};
+	spi_s23: spi-s23 {
+		compatible = "spi-gpio";
+		#address-cells = <0x1>;
+		#size-cells = <0x0>;
+		cs-gpios  = <&gpio1 17 GPIO_ACTIVE_HIGH>; /* S23A gpio1.17 0x44 */
+		gpio-sck  = <&gpio1 18 GPIO_ACTIVE_HIGH>; /* S23B gpio1.18 0x48 */
+		gpio-mosi = <&gpio1 19 GPIO_ACTIVE_HIGH>; /* S23C gpio1.19 0x4c */
+		gpio-miso = <&gpio1 28 GPIO_ACTIVE_HIGH>; /* S23D gpio1.28 0x78 */
+		num-chipselects = <1>;
+		pinctrl-names = "default";
+		pinctrl-0 = <&spi_s23_pins>;
+		status = "disabled";
+		spi@0 {
+			compatible = "tps,spigen";
+			spi-max-frequency = <57600000>;
+			reg = <0>;
+		};
+	};
+
+	twg_s01: twg-s01 {
+		compatible = "tps,wg-gpio";
+		wg-oc = <&gpio0 15 GPIO_ACTIVE_HIGH>; /* S01A gpio0.15 0x184 */
+		wg-w1 = <&gpio0 14 GPIO_ACTIVE_HIGH>; /* S01B gpio0.14 0x180 */
+		wg-sw = <&gpio0 13 GPIO_ACTIVE_HIGH>; /* S01C gpio0.13 0x17c */
+		wg-w0 = <&gpio0 12 GPIO_ACTIVE_HIGH>; /* S01D gpio0.12 0x178 */
+		pinctrl-names = "default";
+		pinctrl-0 = <&i2c_s01_pins>;
+		status = "disabled";
+	};
+	twg_s03: twg-s03 {
+		compatible = "tps,wg-gpio";
+		wg-oc = <&gpio0 19 GPIO_ACTIVE_HIGH>; /* S03A gpio0.19 0x1b0 */
+		wg-w1 = <&gpio0  7 GPIO_ACTIVE_HIGH>; /* S03B gpio0.07 0x164 */
+		wg-sw = <&gpio3 21 GPIO_ACTIVE_HIGH>; /* S03C gpio3.21 0x1ac */
+		wg-w0 = <&gpio3 13 GPIO_ACTIVE_HIGH>; /* S03D gpio3.13 0x234 */
+		pinctrl-names = "default";
+		pinctrl-0 = <&i2c_s03_pins>;
+		status = "disabled";
+	};
+	twg_s05: twg-s05 {
+		compatible = "tps,wg-gpio";
+		wg-oc = <&gpio2 29 GPIO_ACTIVE_HIGH>; /* S05A gpio2.29 0xfc */
+		wg-w1 = <&gpio2 28 GPIO_ACTIVE_HIGH>; /* S05B gpio2.28 0xf8 */
+		wg-sw = <&gpio2 17 GPIO_ACTIVE_HIGH>; /* S05C gpio2.17 0xcc */
+		wg-w0 = <&gpio2 16 GPIO_ACTIVE_HIGH>; /* S05D gpio2.16 0xc8 */
+		pinctrl-names = "default";
+		pinctrl-0 = <&i2c_s05_pins>;
+		status = "disabled";
+	};
+	twg_s07: twg-s07 {
+		compatible = "tps,wg-gpio";
+		wg-oc = <&gpio2 27 GPIO_ACTIVE_HIGH>; /* S07A gpio2.27 0xf4 */
+		wg-w1 = <&gpio0 20 GPIO_ACTIVE_HIGH>; /* S07B gpio0.20 0x1b4 */
+		wg-sw = <&gpio0 29 GPIO_ACTIVE_HIGH>; /* S07C gpio0.29 0x144 */
+		wg-w0 = <&gpio1 20 GPIO_ACTIVE_HIGH>; /* S07D gpio1.20 0x50 */
+		pinctrl-names = "default";
+		pinctrl-0 = <&i2c_s07_pins>;
+		status = "disabled";
+	};
+	twg_s09: twg-s09 {
+		compatible = "tps,wg-gpio";
+		wg-oc = <&gpio2 14 GPIO_ACTIVE_HIGH>; /* S09A gpio2.14 0xc0 */
+		wg-w1 = <&gpio2 15 GPIO_ACTIVE_HIGH>; /* S09B gpio2.15 0xc4 */
+		wg-sw = <&gpio0 11 GPIO_ACTIVE_HIGH>; /* S09C gpio0.11 0xdc */
+		wg-w0 = <&gpio0 10 GPIO_ACTIVE_HIGH>; /* S09D gpio0.10 0xd8 */
+		pinctrl-names = "default";
+		pinctrl-0 = <&i2c_s09_pins>;
+		status = "disabled";
+	};
+	twg_s11: twg-s11 {
+		compatible = "tps,wg-gpio";
+		wg-oc = <&gpio2 26 GPIO_ACTIVE_HIGH>; /* S11A gpio2.26 0xf0 */
+		wg-w1 = <&gpio1 23 GPIO_ACTIVE_HIGH>; /* S11B gpio1.23 0x5c */
+		wg-sw = <&gpio1 26 GPIO_ACTIVE_HIGH>; /* S11C gpio1.26 0x68 */
+		wg-w0 = <&gpio3 18 GPIO_ACTIVE_HIGH>; /* S11D gpio3.18 0x1a0 */
+		pinctrl-names = "default";
+		pinctrl-0 = <&i2c_s11_pins>;
+		status = "disabled";
+	};
+	twg_s13: twg-s13 {
+		compatible = "tps,wg-gpio";
+		wg-oc = <&gpio1  9 GPIO_ACTIVE_HIGH>; /* S13A gpio1.09 0x16c */
+		wg-w1 = <&gpio1  8 GPIO_ACTIVE_HIGH>; /* S13B gpio1.08 0x168 */
+		wg-sw = <&gpio0  9 GPIO_ACTIVE_HIGH>; /* S13C gpio0.09 0xd5 */
+		wg-w0 = <&gpio0  8 GPIO_ACTIVE_HIGH>; /* S13D gpio0.08 0xd0 */
+		pinctrl-names = "default";
+		pinctrl-0 = <&i2c_s13_pins>;
+		status = "disabled";
+	};
+	twg_s15: twg-s15 {
+		compatible = "tps,wg-gpio";
+		wg-oc = <&gpio3  6 GPIO_ACTIVE_HIGH>; /* S15A gpio3.06 0x18c */
+		wg-w1 = <&gpio3  5 GPIO_ACTIVE_HIGH>; /* S15B gpio3.05 0x188 */
+		wg-sw = <&gpio3 14 GPIO_ACTIVE_HIGH>; /* S15C gpio3.14 0x190 */
+		wg-w0 = <&gpio3 17 GPIO_ACTIVE_HIGH>; /* S15D gpio3.17 0x19c */
+		pinctrl-names = "default";
+		pinctrl-0 = <&i2c_s15_pins>;
+		status = "disabled";
+	};
+	twg_s17: twg-s17 {
+		compatible = "tps,wg-gpio";
+		wg-oc = <&gpio1 11 GPIO_ACTIVE_HIGH>; /* S17A gpio1.11 0x174 */
+		wg-w1 = <&gpio1 10 GPIO_ACTIVE_HIGH>; /* S17B gpio1.10 0x170 */
+		wg-sw = <&gpio2 25 GPIO_ACTIVE_HIGH>; /* S17C gpio2.25 0xec */
+		wg-w0 = <&gpio2 23 GPIO_ACTIVE_HIGH>; /* S17D gpio2.23 0xe4 */
+		pinctrl-names = "default";
+		pinctrl-0 = <&i2c_s17_pins>;
+		status = "disabled";
+	};
+	twg_s19: twg-s19 {
+		compatible = "tps,wg-gpio";
+		wg-oc = <&gpio2 31 GPIO_ACTIVE_HIGH>; /* S19A gpio2.31 0x104 */
+		wg-w1 = <&gpio2 30 GPIO_ACTIVE_HIGH>; /* S19B gpio2.30 0x100 */
+		wg-sw = <&gpio1 31 GPIO_ACTIVE_HIGH>; /* S19C gpio1.31 0x84 */
+		wg-w0 = <&gpio1 30 GPIO_ACTIVE_HIGH>; /* S19D gpio1.30 0x80 */
+		pinctrl-names = "default";
+		pinctrl-0 = <&i2c_s19_pins>;
+		status = "disabled";
+	};
+	twg_s21: twg-s21 {
+		compatible = "tps,wg-gpio";
+		wg-oc = <&gpio2  1 GPIO_ACTIVE_HIGH>; /* S21A gpio2.01 0x8c */
+		wg-w1 = <&gpio2  0 GPIO_ACTIVE_HIGH>; /* S21B gpio2.00 0x88 */
+		wg-sw = <&gpio3 16 GPIO_ACTIVE_HIGH>; /* S21C gpio3.16 0x198 */
+		wg-w0 = <&gpio0  5 GPIO_ACTIVE_HIGH>; /* S21D gpio0.05 0x15c */
+		pinctrl-names = "default";
+		pinctrl-0 = <&i2c_s21_pins>;
+		status = "disabled";
+	};
+	twg_s23: twg-s23 {
+		compatible = "tps,wg-gpio";
+		wg-oc = <&gpio1 17 GPIO_ACTIVE_HIGH>; /* S23A gpio1.17 0x44 */
+		wg-w1 = <&gpio1 18 GPIO_ACTIVE_HIGH>; /* S23B gpio1.18 0x48 */
+		wg-sw = <&gpio1 19 GPIO_ACTIVE_HIGH>; /* S23C gpio1.19 0x4c */
+		wg-w0 = <&gpio1 28 GPIO_ACTIVE_HIGH>; /* S23D gpio1.28 0x78 */
+		pinctrl-names = "default";
+		pinctrl-0 = <&i2c_s23_pins>;
+		status = "disabled";
+	};
+
+};
+
+/* HW i2c on S15 */
+&i2c0 {
+	status = "disabled";
+	pinctrl-names = "default";
+	pinctrl-0 = <&i2c0_pins>;
+};
+
+&uart0 {
+	pinctrl-names = "default", "norts";
+	pinctrl-0 = <&uart0_pins>;
+	pinctrl-1 = <&uart0_norts_pins>;
+	rts-gpio = <&gpio2 25 GPIO_ACTIVE_HIGH>;
+	rs485-rts-delay = <0 0>;
+	linux,rs485-enabled-at-boot-time;
+	rs485-rts-active-high;
+	rs485-rx-during-tx;
+	status = "disabled";
+};
+&uart1 {
+	pinctrl-names = "default", "norts";
+	pinctrl-0 = <&uart1_pins/* &uart1_pins_485*/>;
+	pinctrl-1 = <&uart1_norts_pins/* &uart1_pins_485*/>;
+	rts-gpio = <&gpio0 13 GPIO_ACTIVE_HIGH>;
+	rs485-rts-delay = <0 0>;
+	linux,rs485-enabled-at-boot-time;
+	rs485-rts-active-high;
+	rs485-rx-during-tx;
+	status = "disabled";
+};
+
+&uart3 {
+	pinctrl-names = "default", "norts";
+	pinctrl-0 = <&uart3_pins/* &uart3_pins_485*/>;
+	pinctrl-1 = <&uart3_norts_pins/* &uart3_pins_485*/>;
+	rts-gpio = <&gpio2 17 GPIO_ACTIVE_HIGH>;
+	rs485-rts-delay = <0 0>;
+	linux,rs485-enabled-at-boot-time;
+	rs485-rts-active-high;
+	rs485-rx-during-tx;
+	status = "disabled";
+};
+&uart4 {
+	pinctrl-names = "default", "norts";
+	pinctrl-0 = <&uart4_pins/* &uart4_pins_485*/>;
+	pinctrl-1 = <&uart4_norts_pins/* &uart4_pins_485*/>;
+	rts-gpio = <&gpio0 9 GPIO_ACTIVE_HIGH>;
+	rs485-rts-delay = <0 0>;
+	linux,rs485-enabled-at-boot-time;
+	rs485-rts-active-high;
+	rs485-rx-during-tx;
+	status = "disabled";
+};
+&uart5 {
+	pinctrl-names = "default", "norts";
+	pinctrl-0 = <&uart5_pins/* &uart5_pins_485*/>;
+	pinctrl-1 = <&uart5_norts_pins/* &uart5_pins_485*/>;
+	rts-gpio = <&gpio0 11 GPIO_ACTIVE_HIGH>;
+	rs485-rts-delay = <0 0>;
+	linux,rs485-enabled-at-boot-time;
+	rs485-rts-active-high;
+	rs485-rx-during-tx;
+	status = "disabled";
+};
+
+&dcan0 { /* S17 */
+	pinctrl-names = "default";
+	pinctrl-0 = <&dcan0_pins>;
+	status = "disabled";
+};
+&dcan1 { /* S19 */
+	pinctrl-names = "default";
+	pinctrl-0 = <&dcan1_pins>;
+	status = "disabled";
+};
+
+&mcasp0 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&sound_pins>;
+	status = "disabled";
+
+	op-mode = <0>;          /* MCASP_IIS_MODE */
+	tdm-slots = <2>;
+	/* 4 serializers */
+	serial-dir = <  /* 0: INACTIVE, 1: TX, 2: RX */
+		2 0 1 0
+	>;
+	tx-num-evt = <32>;
+	rx-num-evt = <32>;
+};
+