--- linux/arch/arm/boot/dts/am335x-KLM601.dts	1970-01-01 03:00:00.000000000 +0300
+++ linux/arch/arm/boot/dts/am335x-KLM601.dts	2014-12-16 10:12:59.961528417 +0300
@@ -0,0 +1,123 @@
+/*
+ * KLM601
+ * Copyright (C) 2016 Tibbo Tech. - http://tibbo.com/tps/
+ * License: GPLv2 or later
+ */
+/dts-v1/;
+
+#include "am33xx.tpp.dtsi"
+#include "am335x-TPP.dtsi"
+#include "am335x-TPPs3.dtsi"
+
+/ {
+	model = "AM335x KLM601 test";
+	compatible = "ti,am335x-tpp3", "ti,am335x-tpp", "ti,am33xx";
+
+/*
+	lcd_bl_gpio: lcd_bl_gpio {
+		compatible = "gpio-backlight";
+		pinctrl-names = "default";
+		pinctrl-0 = <&lcd_backlight_pins>;
+		gpios = <&gpio3 17 GPIO_ACTIVE_HIGH>;
+		default-on;
+	};
+*/
+	lcd_bl_pwm: lcd_bl_pwm {
+		compatible = "pwm-backlight";
+		pwms = <&ecap2 0 50000 0>;
+		brightness-levels = <0 51 53 56 62 75 101 152 255>;
+		default-brightness-level = <8>;
+	};
+
+	panel {
+		compatible = "ti,tilcdc,panel";
+		pinctrl-names = "default";
+		pinctrl-0 = <&hdmi_pins>;
+		/* backlight = <&lcd_bl_gpio>; */
+		backlight = <&lcd_bl_pwm>;
+		status = "okay";
+		panel-info {
+			ac-bias           = <255>;
+			ac-bias-intrpt    = <0>;
+			dma-burst-sz      = <16>;
+			bpp               = <16>;
+			fdd               = <0x80>;
+			sync-edge         = <1>;
+			sync-ctrl         = <1>;
+			raster-order      = <0>;
+			fifo-th           = <0>;
+		};
+		display-timings {
+			native-mode = <&timing0>;
+			timing0: 320x240 {
+				hactive         = <320>;
+				vactive         = <240>;
+				hback-porch     = <39>;		/* pixels */
+				hfront-porch    = <19>;		/* pixels */
+				hsync-len       = <30>;		/* pixels */
+				vback-porch     = <16>;		/* lines */
+				vfront-porch    = <3>;		/* lines */
+				vsync-len       = <3>;		/* lines */
+				clock-frequency = <6500000>;
+				hsync-active    = <0>;
+				vsync-active    = <0>;
+				de-active       = <1>;
+			};
+		};
+	};
+
+};
+
+&am33xx_pinmux {
+/*	pinctrl-0 =<&lcd_backlight_pins>;*/
+	lcd_backlight_pins: lcd_backlight_pins {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x19c, (PIN_OUTPUT | MUX_MODE4/*7*/))	/* gpio3.17 pin113 M4=pwm2/M7=gpio */
+		>;
+	};
+	spi_wifi_add_gpio: spi_wifi_add_gpio {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x988, (PIN_OUTPUT | MUX_MODE7))	/* v-v? gpio3.5 pin116 wifi_reset */
+			AM33XX_IOPAD(0x98c, (PIN_INPUT | MUX_MODE7))	/* v-v? gpio3.6 pin115 wifi_irq (extra1) */
+		>;
+	};
+	mmc2_pins_g: mmc2_pins_g {
+		pinctrl-single,pins = <
+			AM33XX_IOPAD(0x844, (PIN_INPUT_PULLUP    | MUX_MODE7))	/* mmc2_d0  S23_A */
+			AM33XX_IOPAD(0x848, (PIN_INPUT_PULLUP    | MUX_MODE7))	/* mmc2_d1  S23_B */
+			AM33XX_IOPAD(0x84c, (PIN_INPUT_PULLUP    | MUX_MODE7))	/* mmc2_d2  S23_C */
+			AM33XX_IOPAD(0x878, (PIN_INPUT_PULLUP    | MUX_MODE7))	/* mmc2_d3  S23_D */
+			AM33XX_IOPAD(0x888, (PIN_INPUT_PULLUP    | MUX_MODE7))	/* mmc2_cmd S21_B */
+			AM33XX_IOPAD(0x88c, (PIN_INPUT_PULLUP    | MUX_MODE7))	/* mmc2_clk S21_A */
+			AM33XX_IOPAD(0x198, (PIN_INPUT | MUX_MODE7))	/* mmc2_cd  S21_C */
+		>;
+	};
+};
+&mmc3 {  status = "okay";  };
+&cppi41dma {  status = "okay";  };
+&usb {  status = "okay";  };
+&usb_ctrl_mod {  status = "okay";  };
+&usb0_phy {  status = "okay";  };
+&usb0 {  status = "okay";  };
+&usb1_phy {  status = "okay";  };
+&usb1 {  status = "okay";  };
+&uart0 {  status = "okay";  };
+/*&uart3 {  status = "okay";  };*/
+/*&uart4 {  status = "okay";  };*/
+&spi0 {  status = "okay";  };
+&ga1000v0 {
+ status = "okay";
+/* gpio-irq = <102>;*/	/* wifi-ext1, gpio3.6, 0x18c */
+/* gpio-reset = <&gpio3 5 GPIO_ACTIVE_HIGH>;*/ /* wifi_rst gpio3.5 0x188 */
+};
+&cpsw_emac0 {  status = "disabled";  };
+&mac {  status = "disabled";  };
+&lcdc {  status = "okay";  };
+&epwmss2 {  status = "okay";  };
+/* pwm backlite */
+&ecap2 {
+ status = "okay";
+ pinctrl-names = "default";
+ pinctrl-0 = <&lcd_backlight_pins>;
+};
+/* pwm backlite / */
+