--- linux/arch/arm/boot/dts/am335x-TPPs3-t-i2c.dts	1970-01-01 03:00:00.000000000 +0300
+++ linux/arch/arm/boot/dts/am335x-TPPs3-t-i2c.dts	2014-12-16 10:12:59.961528417 +0300
@@ -0,0 +1,28 @@
+/*
+ * Tibbo Project PCB 3 for Linux (TPP3)
+ * Copyright (C) 2016 Tibbo Tech. - http://tibbo.com/tps/
+ * License: GPLv2 or later
+ */
+/dts-v1/;
+
+#include "am33xx.tpp.dtsi"
+#include "am335x-TPP.dtsi"
+#include "am335x-TPPs3.dtsi"
+
+/ {
+	model = "AM335x TPPs3 I2C";
+	compatible = "ti,am335x-tpp3", "ti,am335x-tpp", "ti,am33xx";
+
+	spi_s05 {
+		pinctrl-names = "default";
+		pinctrl-0 = <&spi_s05_pins>;
+	};
+
+};
+
+&uart0 {  status = "okay";  };  /* S17, dbg */
+&uart4 {  status = "okay";  };  /* S13 */
+&uart5 {  status = "okay";  };  /* S09 */
+&i2c_s01 {  status = "okay";  };  /* S01 */
+&i2c_s03 {  status = "okay";  };  /* S03 */
+/*&spi_s05 {  status = "okay";  };*/  /* S05 */
+