--- linux/arch/arm/boot/dts/am335x-TPPs3-t-uart.dts	1970-01-01 03:00:00.000000000 +0300
+++ linux/arch/arm/boot/dts/am335x-TPPs3-t-uart.dts	2014-12-16 10:12:59.961528417 +0300
@@ -0,0 +1,21 @@
+/*
+ * Tibbo Project PCB 3 for Linux (TPP3)
+ * Copyright (C) 2016 Tibbo Tech. - http://tibbo.com/tps/
+ * License: GPLv2 or later
+ */
+/dts-v1/;
+
+#include "am33xx.tpp.dtsi"
+#include "am335x-TPP.dtsi"
+#include "am335x-TPPs3.dtsi"
+
+/ {
+	model = "AM335x TPPs3 UARTs";
+	compatible = "ti,am335x-tpp3", "ti,am335x-tpp", "ti,am33xx";
+};
+
+&uart0 {  status = "okay";  };  /* S17, dbg */
+&uart1 {  status = "okay";  };  /* S01[+S03] */
+&uart3 {  status = "okay";  };  /* S05[+S07] */
+&uart4 {  status = "okay";  };  /* S13[+S15] */
+&uart5 {  status = "okay";  };  /* S09[+S11] */
+