--- linux/arch/arm/boot/dts/am335x-TPPs3.dts	1970-01-01 03:00:00.000000000 +0300
+++ linux/arch/arm/boot/dts/am335x-TPPs3.dts	2014-12-16 10:12:59.961528417 +0300
@@ -0,0 +1,39 @@
+/*
+ * Tibbo Project PCB 3 for Linux (TPP3)
+ * Copyright (C) 2016 Tibbo Tech. - http://tibbo.com/tps/
+ * License: GPLv2 or later
+ */
+/dts-v1/;
+
+#include "am33xx.tpp.dtsi"
+#include "am335x-TPP.dtsi"
+#include "am335x-TPPs3.dtsi"
+
+/ {
+	model = "AM335x TPPs3 GPIOs";
+	compatible = "ti,am335x-tpp3", "ti,am335x-tpp", "ti,am33xx";
+};
+
+&am33xx_pinmux {
+	pinctrl-0 = <&gpio_pins>;
+	gpio_pins: gpio_pins {
+		pinctrl-single,pins = <
+			/*0x88  (PIN_OUTPUT | MUX_MODE7)*/	/* gpio2.00 pin64  S19C */
+			/*0xf0  (PIN_OUTPUT | MUX_MODE7)*/	/* gpio2.26 pin90  S11A */
+			/*0x1b0 (PIN_OUTPUT | MUX_MODE7)*/	/* gpio0.19 pin19  S3A  */
+			AM33XX_IOPAD(0x878, (PIN_OUTPUT | INPUT_EN | MUX_MODE7))	/* ? gpio1.28 pin60  S23D (mmc2) */
+			AM33XX_IOPAD(0x88c, (PIN_OUTPUT | INPUT_EN | MUX_MODE7))	/* ? gpio2.01 pin65  S21A (mmc2) */
+			AM33XX_IOPAD(0x888, (PIN_OUTPUT | INPUT_EN | MUX_MODE7)) /* ? gpio2.00 pin64  S21B (mmc2) */
+			AM33XX_IOPAD(0x998, (PIN_OUTPUT | INPUT_EN | MUX_MODE7)) /* ? gpio3.16 pin112 S21C (mmc2) */
+			AM33XX_IOPAD(0x844, (PIN_OUTPUT | INPUT_EN | MUX_MODE7)) /* ? gpio1.18 pin50  S23B (mmc2) */
+			AM33XX_IOPAD(0x848, (PIN_OUTPUT | INPUT_EN | MUX_MODE7)) /* ? gpio1.18 pin50  S23B (mmc2) */
+			AM33XX_IOPAD(0x84c, (PIN_OUTPUT | INPUT_EN | MUX_MODE7)) /* ? gpio1.19 pin51  S23C (mmc2) */
+			AM33XX_IOPAD(0xa34, (PIN_OUTPUT | INPUT_EN | MUX_MODE7))	/* ? gpio3.13 pin109 S3D (usb1) */
+			AM33XX_IOPAD(0x85c, (PIN_OUTPUT | INPUT_EN | MUX_MODE7))	/* gpio2.26 pin90 S11A */
+			AM33XX_IOPAD(0x98c, (PIN_OUTPUT | INPUT_EN | MUX_MODE7))	/* gpio2.26 pin90 S15A */
+			AM33XX_IOPAD(0x988, (PIN_OUTPUT | INPUT_EN | MUX_MODE7))	/* gpio2.26 pin90 S15B */
+		>;
+	};
+};
+
+&uart0 {  status = "okay";  };  /* S17, dbg */
+