--- linux/arch/arm/boot/dts/am33xx-clocks.dtsi	2015-03-27 01:17:07.832357366 +0300
+++ linux/arch/arm/boot/dts/am33xx-clocks.dtsi	2015-03-27 01:21:02.013765478 +0300
@@ -372,47 +372,47 @@
 		reg = <0x0528>;
 	};
 
-	timer2_fck: timer2_fck {
-		#clock-cells = <0>;
-		compatible = "ti,mux-clock";
-		clocks = <&tclkin_ck>, <&sys_clkin_ck>, <&clkdiv32k_ick>;
-		reg = <0x0508>;
-	};
-
-	timer3_fck: timer3_fck {
-		#clock-cells = <0>;
-		compatible = "ti,mux-clock";
-		clocks = <&tclkin_ck>, <&sys_clkin_ck>, <&clkdiv32k_ick>;
-		reg = <0x050c>;
-	};
-
-	timer4_fck: timer4_fck {
-		#clock-cells = <0>;
-		compatible = "ti,mux-clock";
-		clocks = <&tclkin_ck>, <&sys_clkin_ck>, <&clkdiv32k_ick>;
-		reg = <0x0510>;
-	};
-
-	timer5_fck: timer5_fck {
-		#clock-cells = <0>;
-		compatible = "ti,mux-clock";
-		clocks = <&tclkin_ck>, <&sys_clkin_ck>, <&clkdiv32k_ick>;
-		reg = <0x0518>;
-	};
-
-	timer6_fck: timer6_fck {
-		#clock-cells = <0>;
-		compatible = "ti,mux-clock";
-		clocks = <&tclkin_ck>, <&sys_clkin_ck>, <&clkdiv32k_ick>;
-		reg = <0x051c>;
-	};
-
-	timer7_fck: timer7_fck {
-		#clock-cells = <0>;
-		compatible = "ti,mux-clock";
-		clocks = <&tclkin_ck>, <&sys_clkin_ck>, <&clkdiv32k_ick>;
-		reg = <0x0504>;
-	};
+/*	timer2_fck: timer2_fck {*/
+/*		#clock-cells = <0>;*/
+/*		compatible = "ti,mux-clock";*/
+/*		clocks = <&tclkin_ck>, <&sys_clkin_ck>, <&clkdiv32k_ick>;*/
+/*		reg = <0x0508>;*/
+/*	};*/
+
+/*	timer3_fck: timer3_fck {*/
+/*		#clock-cells = <0>;*/
+/*		compatible = "ti,mux-clock";*/
+/*		clocks = <&tclkin_ck>, <&sys_clkin_ck>, <&clkdiv32k_ick>;*/
+/*		reg = <0x050c>;*/
+/*	};*/
+
+/*	timer4_fck: timer4_fck {*/
+/*		#clock-cells = <0>;*/
+/*		compatible = "ti,mux-clock";*/
+/*		clocks = <&tclkin_ck>, <&sys_clkin_ck>, <&clkdiv32k_ick>;*/
+/*		reg = <0x0510>;*/
+/*	};*/
+
+/*	timer5_fck: timer5_fck {*/
+/*		#clock-cells = <0>;*/
+/*		compatible = "ti,mux-clock";*/
+/*		clocks = <&tclkin_ck>, <&sys_clkin_ck>, <&clkdiv32k_ick>;*/
+/*		reg = <0x0518>;*/
+/*	};*/
+
+/*	timer6_fck: timer6_fck {*/
+/*		#clock-cells = <0>;*/
+/*		compatible = "ti,mux-clock";*/
+/*		clocks = <&tclkin_ck>, <&sys_clkin_ck>, <&clkdiv32k_ick>;*/
+/*		reg = <0x051c>;*/
+/*	};*/
+
+/*	timer7_fck: timer7_fck {*/
+/*		#clock-cells = <0>;*/
+/*		compatible = "ti,mux-clock";*/
+/*		clocks = <&tclkin_ck>, <&sys_clkin_ck>, <&clkdiv32k_ick>;*/
+/*		reg = <0x0504>;*/
+/*	};*/
 
 	usbotg_fck: usbotg_fck {
 		#clock-cells = <0>;
