--- a/drivers/mtd/nand/Makefile	2017-09-20 09:25:38.000000000 +0300
+++ a/drivers/mtd/nand/Makefile	2018-11-12 06:42:54.000000000 +0300
@@ -59,6 +59,8 @@
 obj-$(CONFIG_MTD_NAND_BRCMNAND)		+= brcmnand/
 obj-$(CONFIG_MTD_NAND_QCOM)		+= qcom_nandc.o
 obj-$(CONFIG_MTD_NAND_MTK)		+= mtk_nand.o mtk_ecc.o
+obj-$(CONFIG_MTD_NAND_SP)		+= sp_spinand.o sp_bch.o
+
 
 nand-objs := nand_base.o nand_bbt.o nand_timings.o nand_ids.o
 nand-objs += nand_amd.o
