--- a/drivers/usb/Kconfig	2017-09-20 09:25:38.000000000 +0300
+++ a/drivers/usb/Kconfig	2018-11-12 06:42:55.000000000 +0300
@@ -2,6 +2,21 @@
 # USB device configuration
 #
 
+config GEMINI_USB
+	bool "Support Gemini USB"
+	select SCSI
+	select USB
+	select USB_SUPPORT
+	select USB_ARCH_HAS_EHCI
+	select USB_ARCH_HAS_OHCI
+#	select USB_EHCI_HCD_PLATFORM
+#	select USB_OHCI_HCD_PLATFORM
+#	select USB_EHCI_HCD
+#	select USB_OHCI_HCD
+	select USB_STORAGE
+	help
+	  Sunplus Gemini platform USB support
+
 config USB_OHCI_BIG_ENDIAN_DESC
 	bool
 
@@ -28,6 +43,22 @@
 	  You will also need drivers from the following menu to make use of it.
 
 if USB_SUPPORT
+# usb ports freely determine enable/disable
+config USB_PORT0
+	boolean "Sunplus Gemini USB PORT 0"
+	default y
+
+config USB_PORT1
+	boolean "Sunplus Gemini USB PORT 1"
+	default y
+
+config USB_GADGET_ENABLED
+	bool
+	default y if (USB_PORT0 || USB_PORT1)
+
+config USB_GADGET_PORT0_ENABLED
+	boolean "Sunplus Gemini USB gadget port 0"
+	default n
 
 config USB_COMMON
 	tristate
@@ -85,6 +116,18 @@
 
 if USB
 
+config USB_HOST_ENUM_RETRY
+	bool "Enable USB Host Enum Retry"
+	depends on USB
+
+config USB_MULTIPLE_RESET_PROBLEM_WORKAROUND
+	bool "Enable USB Multiple Reset Problem Workaround"
+	depends on USB
+
+config USB_HOST_NOT_FINISH_QTD_WHEN_DISC_WORKAROUND
+	bool "Enable Workaround USB Host Not Finish QTD When Disconnect"
+	depends on USB
+
 source "drivers/usb/core/Kconfig"
 
 source "drivers/usb/mon/Kconfig"
@@ -162,6 +205,8 @@
 
 source "drivers/usb/gadget/Kconfig"
 
+source "drivers/usb/bc/Kconfig"
+
 source "drivers/usb/typec/Kconfig"
 
 config USB_LED_TRIG
