--- linux/arch/arm/boot/dts/am335x-TPPs3.dts	1970-01-01 03:00:00.000000000 +0300
+++ linux/arch/arm/boot/dts/am335x-TPPs3.dts	2014-12-16 10:12:59.961528417 +0300
@@ -0,0 +1,22 @@
+/*
+ * Tibbo Project PCB 3 for Linux (TPP)
+ * Copyright (C) 2015 Tibbo Tech. - http://tibbo.com/tps/
+ * License: GPLv2 or later
+ */
+/dts-v1/;
+
+#include "am33xx.tpp.dtsi"
+#include "am335x-TPP.dtsi"
+
+/ {
+	model = "AM335x TPPs3";
+	compatible = "ti,am335x-tpp3", "ti,am335x-tpp", "ti,am33xx";
+};
+
+/*
+&ldo3_reg {
+	regulator-min-microvolt = <1800000>;
+	regulator-max-microvolt = <3300000>;
+	regulator-always-on;
+};
+*/
