--- a/drivers/pwm/Kconfig	2017-09-20 09:25:38.000000000 +0300
+++ a/drivers/pwm/Kconfig	2018-11-12 06:42:55.000000000 +0300
@@ -387,6 +387,15 @@
 	  To compile this driver as a module, choose M here: the module
 	  will be called pwm-samsung.
 
+config PWM_SC7021
+	tristate "SC7021 PWM support"
+	depends on ARCH_PENTAGRAM
+	help
+	  Generic PWM framework driver for SC7021.
+
+	  To compile this driver as a module, choose M here: the module
+	  will be called pwm-sc7021.
+
 config PWM_SPEAR
 	tristate "STMicroelectronics SPEAr PWM support"
 	depends on PLAT_SPEAR
