diff -urN linux/drivers/misc/Kconfig linux/drivers/misc/Kconfig
--- linux/drivers/misc/Kconfig	2015-01-26 16:22:38.000000000 +0300
+++ linux/drivers/misc/Kconfig	2015-01-26 16:31:41.545812498 +0300
@@ -527,4 +527,5 @@
 source "drivers/misc/mic/Kconfig"
 source "drivers/misc/genwqe/Kconfig"
 source "drivers/misc/echo/Kconfig"
+source "drivers/misc/cape/Kconfig"
 endmenu
diff -urN linux/drivers/misc/Makefile linux/drivers/misc/Makefile
--- linux/drivers/misc/Makefile	2015-01-19 17:44:26.514359701 +0300
+++ linux/drivers/misc/Makefile	2015-01-20 01:41:38.382952104 +0300
@@ -55,3 +55,4 @@
 obj-$(CONFIG_GENWQE)		+= genwqe/
 obj-$(CONFIG_GENWQE)		+= echo/
 obj-$(CONFIG_VEXPRESS_SYSCFG)	+= vexpress-syscfg.o
+obj-y				+= cape/
