--- a/utils/key2pub.py	2015-07-24 14:35:38.000000000 +0800
+++ a/utils/key2pub.py	2015-07-24 15:05:25.388894659 +0800
@@ -61,7 +61,7 @@ def print_ssl(output, name, val):
     import struct
     output.write('#include <stdint.h>\n')
     if len(struct.pack('@L', 0)) == 8:
-        return print_ssl_64(output, name, val)
+        return print_ssl_32(output, name, val)
     else:
         return print_ssl_32(output, name, val)
 
