--- a/Makefile	2015-07-24 17:16:31.000000000 +0800
+++ a/Makefile	2015-07-24 17:16:54.923031501 +0800
@@ -116,7 +116,7 @@
 
 $(LIBREG): regdb.h reglib.h reglib.c
 	$(NQ) '  CC  ' $@
-	$(Q)$(CC) $(CFLAGS) $(CPPFLAGS) -o $@ -shared -Wl,-soname,$(LIBREG) $^
+	$(Q)$(CC) $(CFLAGS) $(CPPFLAGS) -o $@ -shared -Wno-unused-variable -Wl,-soname,$(LIBREG) $^
 
 install-libreg-headers:
 	$(NQ) '  INSTALL  libreg-headers'
