--- a/arch/arm/dts/Makefile	2021-10-30 12:45:51.720639152 +0300
+++ a/arch/arm/dts/Makefile	2021-10-30 23:30:32.069081014 +0300
@@ -1035,6 +1035,7 @@
 dtb-$(CONFIG_TARGET_PENTAGRAM_COMMON) += sp7021-dsx.dtb
 dtb-$(CONFIG_TARGET_PENTAGRAM_COMMON) += sp7021-ltpp3g2revD.dtb
 dtb-$(CONFIG_TARGET_PENTAGRAM_COMMON) += sp7021-ltpp3g2revD-uart1.dtb
+dtb-$(CONFIG_TARGET_PENTAGRAM_COMMON) += sp7021-ev.dtb
 
 targets += $(dtb-y)
 
