--- a/Makefile	2019-06-02 18:33:40.094174511 +0300
+++ a/Makefile	2019-06-02 18:35:54.817224750 +0300
@@ -878,7 +878,7 @@
 	@# options are whitelisted, so new ones should not be added.
 	@# create u-boot.img
 	@echo "Wrap u-boot image..."
-	./tools/add_uhdr.sh $(img_name) u-boot.bin u-boot.img 0x200040 0x200040
+	../../git/tools/add_uhdr.sh $(img_name) u-boot.bin u-boot.img 0x200040 0x200040
 	@img_sz=`du -sb u-boot.img | cut -f1` ; \
 	printf "size: %d (hex %x)\n" $$img_sz $$img_sz
 	$(call cmd,cfgcheck,u-boot.cfg)
