--- a/include/configs/pentagram_common.h	2020-05-31 22:16:20.132805536 +0300
+++ a/include/configs/pentagram_common.h	2020-06-01 01:45:08.283528360 +0300
@@ -359,6 +359,11 @@
 	dbg_scr("echo nonosize ${sz_nonos}  addr_dst_nonos ${addr_dst_nonos};")\
 	"mmc read ${addr_dst_nonos} ${addr_src_nonos} ${sz_nonos}; " \
 	"sp_nonos_go ${addr_dst_nonos};"\
+	"mmc read ${addr_tmp_header} ${addr_src_dtb} 0x1; " \
+	"setenv tmpval 0; setexpr tmpaddr ${addr_tmp_header} + 0x4; run be2le; " \
+	"setexpr sz_dtb ${tmpval} + 0x28; " \
+	"setexpr sz_dtb ${sz_dtb} + 0x200; setexpr sz_dtb ${sz_dtb} / 0x200; " \
+	"mmc read ${addr_dst_dtb} ${addr_src_dtb} ${sz_dtb}; " \
 	"mmc read ${addr_tmp_header} ${addr_src_kernel} 0x1; " \
 	"setenv tmpval 0; setexpr tmpaddr ${addr_tmp_header} + 0x0c; run be2le; " \
 	"setexpr sz_kernel ${tmpval} + 0x40; " \
@@ -392,7 +397,7 @@
 	"if itest ${if_use_nfs_rootfs} == 1; then " \
 		"setenv bootargs root=/dev/nfs nfsroot=${nfs_serverip}:${nfs_rootfs_dir} ip=${nfs_clintip}:${nfs_serverip}:${nfs_gatewayip}:${nfs_netmask}::eth0:off rdinit=/linuxrc noinitrd rw console=ttyS0,115200; "\
 	"fi; " \
-	"bootm ${addr_dst_kernel} - ${fdtcontroladdr}; " \
+	"bootm ${addr_dst_kernel} - ${addr_dst_dtb}; " \
 	"\0" \
 "qk_zmem_boot=sp_go ${addr_dst_kernel} ${fdtcontroladdr}\0" \
 "zmem_boot=bootm ${addr_dst_kernel} - ${fdtcontroladdr}\0" \
