--- a/arch/arm/dts/sp7021-ltpp3g2revD.dts	2021-11-08 01:05:35.858628869 +0300
+++ a/arch/arm/dts/sp7021-ltpp3g2revD.dts	2021-11-08 00:54:24.988293394 +0300
@@ -8,5 +8,21 @@
 #include "sp7021-ltpp3g2revD.dtsi"
 
 / {
-	model = "SP7021 LTPP3G2 Rev. D (empty sockets)";
+	model = "SP7021 LTPP3G2 Rev. D (nousb)";
 };
+
+&sp_uphy0 {  status = "disabled";  };
+&sp_uphy1 {  status = "disabled";  };
+
+&sp_ehci0 {  status = "disabled";  };
+&sp_ehci1 {  status = "disabled";  };
+
+&sp_ohci0 {  status = "disabled";  };
+&sp_ohci1 {  status = "disabled";  };
+
+&sp_udc0 {  status = "disabled";  };
+&sp_udc1 {  status = "disabled";  };
+
+&sp_otg0 {  status = "disabled";  };
+&sp_otg1 {  status = "disabled";  };
+
