--- linux/arch/arm/boot/dts/am33xx.dtsi	2014-12-25 14:42:18.000000000 +0300
+++ linux/arch/arm/boot/dts/am33xx.dtsi	2014-12-25 14:43:42.913797344 +0300
@@ -572,7 +572,7 @@
 			ti,hwmods = "epwmss0";
 			#address-cells = <1>;
 			#size-cells = <1>;
-			status = "disabled";
+			status = "ok";
 			ranges = <0x48300100 0x48300100 0x80   /* ECAP */
 				  0x48300180 0x48300180 0x80   /* EQEP */
 				  0x48300200 0x48300200 0x80>; /* EHRPWM */
@@ -582,7 +582,7 @@
 				#pwm-cells = <3>;
 				reg = <0x48300100 0x80>;
 				ti,hwmods = "ecap0";
-				status = "disabled";
+				status = "ok";
 			};
 
 			ehrpwm0: ehrpwm@48300200 {
@@ -590,7 +590,7 @@
 				#pwm-cells = <3>;
 				reg = <0x48300200 0x80>;
 				ti,hwmods = "ehrpwm0";
-				status = "disabled";
+				status = "ok";
 			};
 		};
 
@@ -600,7 +600,7 @@
 			ti,hwmods = "epwmss1";
 			#address-cells = <1>;
 			#size-cells = <1>;
-			status = "disabled";
+			status = "ok";
 			ranges = <0x48302100 0x48302100 0x80   /* ECAP */
 				  0x48302180 0x48302180 0x80   /* EQEP */
 				  0x48302200 0x48302200 0x80>; /* EHRPWM */
@@ -610,7 +610,7 @@
 				#pwm-cells = <3>;
 				reg = <0x48302100 0x80>;
 				ti,hwmods = "ecap1";
-				status = "disabled";
+				status = "ok";
 			};
 
 			ehrpwm1: ehrpwm@48302200 {
@@ -618,7 +618,7 @@
 				#pwm-cells = <3>;
 				reg = <0x48302200 0x80>;
 				ti,hwmods = "ehrpwm1";
-				status = "disabled";
+				status = "ok";
 			};
 		};
 
@@ -628,7 +628,7 @@
 			ti,hwmods = "epwmss2";
 			#address-cells = <1>;
 			#size-cells = <1>;
-			status = "disabled";
+			status = "ok";
 			ranges = <0x48304100 0x48304100 0x80   /* ECAP */
 				  0x48304180 0x48304180 0x80   /* EQEP */
 				  0x48304200 0x48304200 0x80>; /* EHRPWM */
@@ -638,7 +638,7 @@
 				#pwm-cells = <3>;
 				reg = <0x48304100 0x80>;
 				ti,hwmods = "ecap2";
-				status = "disabled";
+				status = "ok";
 			};
 
 			ehrpwm2: ehrpwm@48304200 {
@@ -646,7 +646,7 @@
 				#pwm-cells = <3>;
 				reg = <0x48304200 0x80>;
 				ti,hwmods = "ehrpwm2";
-				status = "disabled";
+				status = "ok";
 			};
 		};
 
