--- linux/arch/arm/boot/dts/am33xx.dtsi	2014-10-31 00:28:46.855919276 +0300
+++ linux/arch/arm/boot/dts/am33xx.dtsi	2014-10-31 00:28:38.524752353 +0300
@@ -687,12 +687,12 @@
 				reg = <0x4a101000 0x100>;
 			};
 
-			cpsw_emac0: slave@4a100200 {
+			cpsw_emac0: slave@4a100300 {
 				/* Filled in by U-Boot */
 				mac-address = [ 00 00 00 00 00 00 ];
 			};
 
-			cpsw_emac1: slave@4a100300 {
+			cpsw_emac1: slave@4a100200 {
 				/* Filled in by U-Boot */
 				mac-address = [ 00 00 00 00 00 00 ];
 			};
