--- a/drivers/staging/iio/meter/ade7758_core.c	2017-09-27 22:58:47.907456234 +0300
+++ a/drivers/staging/iio/meter/ade7758_core.c	2017-09-27 23:10:48.193467883 +0300
@@ -298,9 +298,29 @@
 	ret = ade7758_spi_write_reg_8(dev, ADE7758_OPMODE, val);
 	if (ret < 0)
 		dev_err(dev, "Failed to write opmode reg\n");
+	/* by Dv
+	ret = ade7758_spi_read_reg_8(dev, ADE7758_COMPMODE, &val);
+	if (ret < 0) {
+		dev_err(dev, "Failed to read compmode reg\n");
+		return ret;
+	}
+	ret &= ~BIT(6);
+	ret &= ~BIT(7);
+	ret = ade7758_spi_write_reg_8(dev, ADE7758_COMPMODE, val);
+	if (ret < 0) dev_err(dev, "Failed to write compmode reg\n");
+	 by Dv */
 	return ret;
 }
 
+/* by Dv */
+#define IIO_DEV_ATTR_COMPMODE(_mode, _show, _store, _addr)     \
+	IIO_DEVICE_ATTR(compmode, _mode, _show, _store, _addr)
+
+static IIO_DEV_ATTR_COMPMODE(S_IRUGO,
+		ade7758_read_8bit,
+		NULL,
+		ADE7758_COMPMODE);
+/* by Dv / */
 static IIO_DEV_ATTR_VPEAK(S_IWUSR | S_IRUGO,
 		ade7758_read_8bit,
 		ade7758_write_8bit,
@@ -599,6 +619,9 @@
 	&iio_dev_attr_avahr.dev_attr.attr,
 	&iio_dev_attr_bvahr.dev_attr.attr,
 	&iio_dev_attr_cvahr.dev_attr.attr,
+/* by Dv */
+	&iio_dev_attr_compmode.dev_attr.attr,
+/* by Dv / */
 	&iio_dev_attr_vpeak.dev_attr.attr,
 	&iio_dev_attr_ipeak.dev_attr.attr,
 	&iio_dev_attr_aphcal.dev_attr.attr,
