--- a/drivers/gpu/drm/omapdrm/omap_fbdev.c	2016-12-08 10:28:52.000000000 +0300
+++ a/drivers/gpu/drm/omapdrm/omap_fbdev.c	2016-12-08 10:30:30.827727557 +0300
@@ -112,8 +112,8 @@
 	dma_addr_t paddr;
 	int ret;
 
-	sizes->surface_bpp = 32;
-	sizes->surface_depth = 24;
+	sizes->surface_bpp = 16; /*32;*/
+	sizes->surface_depth = 16; /* 24; */
 
 	DBG("create fbdev: %dx%d@%d (%dx%d)", sizes->surface_width,
 			sizes->surface_height, sizes->surface_bpp,
@@ -283,7 +283,7 @@
 	/* disable all the possible outputs/crtcs before entering KMS mode */
 	drm_helper_disable_unused_functions(dev);
 
-	ret = drm_fb_helper_initial_config(helper, 32);
+	ret = drm_fb_helper_initial_config(helper, /*32*/16);
 	if (ret)
 		goto fini;
 
