--- a/drivers/staging/rtl8723bs/Makefile	2016-11-19 18:44:44.000000000 +0300
+++ a/drivers/staging/rtl8723bs/Makefile	2016-11-22 11:38:41.763122608 +0300
@@ -32,7 +32,7 @@
 ######### Notify SDIO Host Keep Power During Syspend ##########
 CONFIG_RTW_SDIO_PM_KEEP_POWER = y
 ###################### Platform Related #######################
-CONFIG_PLATFORM_I386_PC = y
+CONFIG_PLATFORM_I386_PC = n
 ###############################################################
 
 ########### COMMON  #################################
