--- a/drivers/spi/spi-omap2-mcspi.c	2017-06-05 10:49:29.000000000 +0300
+++ a/drivers/spi/spi-omap2-mcspi.c	2017-06-05 14:17:03.377695566 +0300
@@ -1393,8 +1393,12 @@
 		of_property_read_u32(node, "ti,spi-num-cs", &num_cs);
 		master->num_chipselect = num_cs;
 		master->bus_num = bus_num++;
-		if (of_get_property(node, "ti,pindir-d0-out-d1-in", NULL))
+		if (of_get_property(node, "ti,pindir-d0-out-d1-in", NULL)) {
 			mcspi->pin_dir = MCSPI_PINDIR_D0_OUT_D1_IN;
+  		  dev_dbg(&pdev->dev, "xxx\n");
+                } else {
+  		  dev_dbg(&pdev->dev, "yyy\n");
+                }
 	} else {
 		pdata = dev_get_platdata(&pdev->dev);
 		master->num_chipselect = pdata->num_cs;
