--- a/drivers/gpu/drm/tilcdc/tilcdc_tfp410.c	2016-12-08 15:13:00.000000000 +0300
+++ a/drivers/gpu/drm/tilcdc/tilcdc_tfp410.c	2016-12-08 19:08:09.635352671 +0300
@@ -139,6 +139,7 @@
 
 	ret = drm_encoder_init(dev, encoder, &tfp410_encoder_funcs,
 			DRM_MODE_ENCODER_TMDS);
+printk(KERN_ERR "%s %s:%d ret:%d\n", __FILE__, __FUNCTION__, __LINE__, ret);
 	if (ret < 0)
 		goto fail;
 
@@ -259,6 +260,7 @@
 	connector->doublescan_allowed = 0;
 
 	ret = drm_mode_connector_attach_encoder(connector, encoder);
+printk(KERN_ERR "%s %s:%d ret:%d\n", __FILE__, __FUNCTION__, __LINE__, ret);
 	if (ret)
 		goto fail;
 
@@ -282,14 +284,17 @@
 	struct drm_encoder *encoder;
 	struct drm_connector *connector;
 
+printk(KERN_ERR "%s %s:%d\n", __FILE__, __FUNCTION__, __LINE__);
 	encoder = tfp410_encoder_create(dev, tfp410_mod);
 	if (!encoder)
 		return -ENOMEM;
 
+printk(KERN_ERR "%s %s:%d\n", __FILE__, __FUNCTION__, __LINE__);
 	connector = tfp410_connector_create(dev, tfp410_mod, encoder);
 	if (!connector)
 		return -ENOMEM;
 
+printk(KERN_ERR "%s %s:%d\n", __FILE__, __FUNCTION__, __LINE__);
 	priv->encoders[priv->num_encoders++] = encoder;
 	priv->connectors[priv->num_connectors++] = connector;
 
@@ -331,16 +336,19 @@
 
 	tilcdc_module_init(mod, "tfp410", &tfp410_module_ops);
 
+printk(KERN_ERR "%s %s:%d\n", __FILE__, __FUNCTION__, __LINE__);
 	pinctrl = devm_pinctrl_get_select_default(&pdev->dev);
 	if (IS_ERR(pinctrl))
 		dev_warn(&pdev->dev, "pins are not configured\n");
 
+printk(KERN_ERR "%s %s:%d\n", __FILE__, __FUNCTION__, __LINE__);
 	if (of_property_read_u32(node, "i2c", &i2c_phandle)) {
 		dev_err(&pdev->dev, "could not get i2c bus phandle\n");
 		goto fail;
 	}
 
 	mod->preferred_bpp = dvi_info.bpp;
+printk(KERN_ERR "%s %s:%d bpp:%d\n", __FILE__, __FUNCTION__, __LINE__, dvi_info.bpp);
 
 	i2c_node = of_find_node_by_phandle(i2c_phandle);
 	if (!i2c_node) {
@@ -348,6 +356,7 @@
 		goto fail;
 	}
 
+printk(KERN_ERR "%s %s:%d\n", __FILE__, __FUNCTION__, __LINE__);
 	tfp410_mod->i2c = of_find_i2c_adapter_by_node(i2c_node);
 	if (!tfp410_mod->i2c) {
 		dev_err(&pdev->dev, "could not get i2c\n");
@@ -357,6 +366,7 @@
 
 	of_node_put(i2c_node);
 
+printk(KERN_ERR "%s %s:%d\n", __FILE__, __FUNCTION__, __LINE__);
 	tfp410_mod->gpio = of_get_named_gpio_flags(node, "powerdn-gpio",
 			0, NULL);
 	if (IS_ERR_VALUE(tfp410_mod->gpio)) {
