--- a/drivers/usb/storage/usb.c	2017-09-20 09:25:38.000000000 +0300
+++ a/drivers/usb/storage/usb.c	2018-11-12 06:42:56.000000000 +0300
@@ -87,7 +87,7 @@
 module_param(delay_use, uint, S_IRUGO | S_IWUSR);
 MODULE_PARM_DESC(delay_use, "seconds to delay before using a new device");
 
-static char quirks[128];
+static char quirks[128] = "1567:8902:r,1234:ffff:n";
 module_param_string(quirks, quirks, sizeof(quirks), S_IRUGO | S_IWUSR);
 MODULE_PARM_DESC(quirks, "supplemental list of device IDs and their quirks");
 
