--- a/drivers/pinctrl/sppctl/sppctl.c	2019-09-20 15:15:47.149186936 +0300
+++ a/drivers/pinctrl/sppctl/sppctl.c	2019-09-20 15:49:40.991331790 +0300
@@ -180,6 +180,7 @@
 // print_device_tree_node( np, 0);
  if ( !( p = devm_kzalloc( &( _pd->dev), sizeof( *p), GFP_KERNEL))) return( -ENOMEM);
  memset( p->name, 0, SPPCTL_MAX_NAM);
+ p->debug = 2;
  if ( np) strcpy( p->name, np->name);
  else strcpy( p->name, MNAME);
  dev_set_name( &( _pd->dev), "%s", p->name);
