--- a/Makefile	2020-05-11 20:23:51.706697362 +0300
+++ a/Makefile	2020-05-11 20:27:45.786695129 +0300
@@ -119,7 +119,7 @@
 	@$(OBJDUMP) -d -S $< > $@
 
 $(BIN)/$(TARGET): $(OBJS) $(LD_FILE)
-	@$(LD) $(OBJS) -o $(BIN)/$(TARGET) -Map $(BIN)/$(TARGET).map $(LDFLAGS) $(LDFLAGS_COM)
+	@$(LD) -fno-stack-protector -shared $(OBJS) -o $(BIN)/$(TARGET) -Map $(BIN)/$(TARGET).map $(LDFLAGS) $(LDFLAGS_COM)
 
 $(LD_FILE): $(LD_SRC) $(OBJS)
 	@$(CPP) -P $(CFLAGS) -x c $(LD_SRC) -o $(LD_FILE)
