--- a/drivers/mmc/host/Kconfig	2017-09-20 09:25:38.000000000 +0300
+++ a/drivers/mmc/host/Kconfig	2018-11-12 06:42:54.000000000 +0300
@@ -3,6 +3,22 @@
 #
 
 comment "MMC/SD/SDIO Host Controller Drivers"
+config SP_SDV2
+	tristate "Sunplus Gemini SoC SD Host Controller support"
+	default y
+	help
+	  If you say yes here, you will get support for the Secure Digital Host
+	  Interface on Sunplus SoCs. This includes support for SD cards.
+
+	  If unsure, say N.
+
+config SP_EMMC
+	bool "Sunplus Gemini eMMC Host Controller v4.51 support"
+	default y
+	help
+	  If you say yes here, you will get support for eMMC host interface 
+	  on sunplus Socs.
+	  If unsure, say N.
 
 config MMC_ARMMMCI
 	tristate "ARM AMBA Multimedia Card Interface support"
