--- a/drivers/mmc/host/Makefile	2017-09-20 09:25:38.000000000 +0300
+++ a/drivers/mmc/host/Makefile	2018-11-12 06:42:54.000000000 +0300
@@ -84,6 +84,8 @@
 obj-$(CONFIG_MMC_SDHCI_ST)		+= sdhci-st.o
 obj-$(CONFIG_MMC_SDHCI_MICROCHIP_PIC32)	+= sdhci-pic32.o
 obj-$(CONFIG_MMC_SDHCI_BRCMSTB)		+= sdhci-brcmstb.o
+obj-$(CONFIG_SP_EMMC)				+= spemmc.o
+obj-$(CONFIG_SP_SDV2)				+= spsdv2.o
 
 ifeq ($(CONFIG_CB710_DEBUG),y)
 	CFLAGS-cb710-mmc	+= -DDEBUG
