--- ./drivers/i2c/algos/i2c-algo-bit.c	2015-04-30 12:08:21.000000000 +0300
+++ ./drivers/i2c/algos/i2c-algo-bit.c	2015-04-30 12:44:47.782714273 +0300
@@ -46,7 +46,7 @@
 MODULE_PARM_DESC(bit_test, "lines testing - 0 off; 1 report; 2 fail if stuck");
 
 #ifdef DEBUG
-static int i2c_debug = 1;
+static int i2c_debug = 2;
 module_param(i2c_debug, int, S_IRUGO | S_IWUSR);
 MODULE_PARM_DESC(i2c_debug,
 		 "debug level - 0 off; 1 normal; 2 verbose; 3 very verbose");
