--- a/drivers/misc/Kconfig	2017-09-20 09:25:38.000000000 +0300
+++ a/drivers/misc/Kconfig	2018-11-12 06:42:54.000000000 +0300
@@ -4,6 +4,38 @@
 
 menu "Misc devices"
 
+config SP_IPC
+	tristate "Sunplus Mailbox (IPC) driver"
+	depends on ARCH_PENTAGRAM
+	default n
+
+config SP_IPC_TEST
+	bool "Sunplus Mailbox (IPC) test"
+	depends on SP_IPC
+	help
+	  echo 0 > /sys/module/sp_ipc/parameters/test
+
+config SP_ICM
+	tristate "Sunplus Input Capture (ICM) driver"
+	depends on ARCH_PENTAGRAM
+	default n
+
+config SP_ICM_TEST
+	bool "Sunplus Input Capture (ICM) test"
+	depends on SP_ICM
+	help
+	  echo h > /sys/module/sp_icm/parameters/test
+
+config SP_DMA0
+	tristate "Sunplus DMA0 driver"
+	depends on ARCH_PENTAGRAM
+
+config SP_DMA0_TEST
+	bool "Sunplus DMA0 test"
+	depends on SP_DMA0
+	help
+	  echo h > /sys/module/sp_dma0/parameters/test
+
 config SENSORS_LIS3LV02D
 	tristate
 	depends on INPUT
