--- a/drivers/net/can/spi/mcp25xxfd/mcp25xxfd_base.c	2020-10-04 19:38:49.048965124 +0300
+++ a/drivers/net/can/spi/mcp25xxfd/mcp25xxfd_base.c	2020-10-04 20:09:08.179947776 +0300
@@ -75,7 +75,7 @@
 	int ret;
 
 	/* as irq_create_fwspec_mapping() can return 0, check for it */
-	if (spi->irq <= 0) {
+	if (spi->irq <= 0 && 1 == 0) {
 		dev_err(&spi->dev, "no valid irq line defined: irq = %i\n",
 			spi->irq);
 		return -EINVAL;
